---
title: "Read plots"
author: "G$"
date: "9/3/2020"
output: pdf_document
---

---
```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(Biostrings)
library(ggplot2)
library(png)
library(cowplot)
options(scipen=999)
library(zoo)
options(knitr.duplicate.label = 'allow')
source("/home/isac/Code/ilee/plot/ilee_plot_utils.R")
library("ggsci")
source("/home/isac/Code/nanopore-methylation-utilities/methylation_R_utils.R")
```

```{r mbed}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/censat/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"

censat = read_tsv(paste0(dat, "/annotations/t2t-chm13.v1.0.cenSat_annotation.bed"), col_names = F) %>%
  mutate(name = ifelse(grepl("HSat1", X4), "HSAT1", X4)) %>%
  mutate(name = ifelse(grepl("HSat2", X4), "HSAT2", name)) %>%
  mutate(name = ifelse(grepl("HSat3", X4), "HSAT3", name)) %>%
  mutate(name = ifelse(grepl("HSat4", X4), "HSAT4", name)) %>%
  mutate(name = ifelse(grepl("HSat5", X4), "HSAT5", name)) %>%
  mutate(name = ifelse(grepl("HSat6", X4), "HSAT6", name)) %>%
  mutate(name = ifelse(grepl("ct", X4), "CT", name)) %>%
  mutate(name = ifelse(grepl("bsat", X4), "BSAT", name)) %>%
  mutate(name = ifelse(grepl("hor", X4), "HOR", name)) %>%
  mutate(name = ifelse(grepl("mon", X4), "MON", name)) %>%
  mutate(name = ifelse(grepl("Alu", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("SATR", X4), "SATR", name)) %>%
  mutate(name = ifelse(grepl("ACRO", X4), "ACRO", name)) %>%
  mutate(name = ifelse(grepl("GSATII", X4), "GSATII", name)) %>%
  mutate(name = ifelse(grepl("TAR", X4), "TAR", name)) %>%
  mutate(name = ifelse(grepl("TE", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("MER", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("MST", X4), "MST", name)) %>%
  mutate(name = ifelse(grepl("CER", X4), "CER", name)) %>%
  mutate(name = ifelse(grepl("L1", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("SST", X4), "SST", name)) %>%
  mutate(name = ifelse(grepl("LSAU", X4), "LSAU", name)) %>%
  mutate(name = ifelse(grepl("GSAT", X4), "GSAT", name)) %>%
  mutate(name = ifelse(grepl("MSAT", X4), "MSAT", name)) %>%
  mutate(name = ifelse(grepl("novel", X4), "novel", name)) %>%
  mutate(name = ifelse(grepl("HERV", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("LT", X4), "TE", name)) %>%
  dplyr::select(c(X1, X2, X3, name)) %>%
  dplyr::rename("chrom" =1, "start" = 2 ,"end" =3) %>%
  group_by(name) %>%
  filter(n() >= 3) %>%
  ungroup()
table(censat$name)
censat.gr <- GRanges(as.data.frame(censat))

repeatColors =c("(CATTC)n" = "#E87C71",
    "(GAATC)n"="#E28455",
    "HOR"="#D78C32",
    "BSAT"="#E370AB",
    "CER" = "#CE9334",
    "HSAT2"="#C19935",
    "HSAT1"="#A2A638",
    "HSAT3"="#8CAC3E",
    "Low_complexity"="#75B042",
    "LSAU"="#54B346",
    "LTR"="#51B756",
    "MST"="#53BB73",
    "GSAT"="#55BE8D",
    "RNA"="#54C0A5",
    "rRNA"="#52BEBB",
    "SAR"="#51BDCE",
    "ACRO"="#4EB8DF",
    "HSAT4"="#53B0E3",
    "SATR"="#5AA5DA",
    "CT"="#6B9AD2",
    "Simple_repeat"="#8992C8",
    "SINE"="#9A8AC1",
    "MON"="#A885BC",
    "SST"="#C378B2",
    "HSAT5"="#ED72A5",
    "HSAT6"="#EF768C", 
    "gap-rDNA"="#ff4000",
    "TE" = "#ffbf00", 
   "TAR"= "#0080ff" )

```


```{r plot, echo=F}
chr="chr19"

reads <- tabix_mbed(paste0(dat, "/censat/", "chr19sst1.mbed"),extcol = "motif",by = "read") 
#chrX:48,626,035-49,040,073



size_sel <- reads %>%
  mutate(rlen = end-start) %>%
  filter(rlen >= 50000) %>%
  mutate(motif = "CG")

cgcalls <- mbedByCall(size_sel) %>%
  drop_na(mcall)

#38800000       39000000
region_start=38850000
region_end= 38900000
reg = cgcalls %>%
  dplyr::filter(start >= region_start) %>%
  dplyr::filter(end <= region_end)

cpg_runs <-getRuns(reg, maxGap = 50)
 

cpg_runs.ordered <- order_reads(cpg_runs)

  cpg_runs <- cpg_runs.ordered$x %>%
    mutate(m = ifelse(values == 1, "Methylated","Unmethylated")) %>%
    mutate(mod = "CpG")
  
pal <- pal_npg("nrc")(10)
meth_pal <- c(pal[8],pal[4]) 
#ids <- order(sapply(seq_along(gcruns.list),function(i){length(unique(gcruns.list$qname))}))

library(adagio)

g <- ggplot(cpg_runs,aes(xmin = start, xmax = end, ymin = ymin, ymax = ymax)) + geom_rect(data = cpg_runs.ordered$bounds, fill = "grey80") + 
    geom_rect(aes(fill = m))  +
#    geom_vline(xintercept = 127638255, linetype == "dashed") +
    scale_fill_manual(name = "State", values = meth_pal) + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), panel.spacing = unit(2, "lines"))
  
  print(g)
ggsave(
  paste0(figs, "/chr19_SST1_SingleRead.pdf"),
  plot = g,
  scale = 1,
  width = 10,
  height = 8,
  dpi = 50
)
```


```{r plot, echo=F}
chr="chr4"

reads <- tabix_mbed(paste0(dat, "/censat/", "chr21_sst1.mbed"),extcol = "motif",by = "read") 
#chrX:48,626,035-49,040,073



size_sel <- reads %>%
  mutate(rlen = end-start) %>%
  filter(rlen >= 50000) %>%
  mutate(motif = "CG")

cgcalls <- mbedByCall(size_sel) %>%
  drop_na(mcall)

#38800000       39000000
region_start=10120000
region_end= 10140000
reg = cgcalls %>%
  dplyr::filter(start >= region_start) %>%
  dplyr::filter(end <= region_end)

cpg_runs <-getRuns(reg, maxGap = 50)
 

cpg_runs.ordered <- order_reads(cpg_runs)

  cpg_runs <- cpg_runs.ordered$x %>%
    mutate(m = ifelse(values == 1, "Methylated","Unmethylated")) %>%
    mutate(mod = "CpG")
  
pal <- pal_npg("nrc")(10)
meth_pal <- c(pal[8],pal[4]) 
#ids <- order(sapply(seq_along(gcruns.list),function(i){length(unique(gcruns.list$qname))}))

library(adagio)

g <- ggplot(cpg_runs,aes(xmin = start, xmax = end, ymin = ymin, ymax = ymax)) + geom_rect(data = cpg_runs.ordered$bounds, fill = "grey80") + 
    geom_rect(aes(fill = m))  +
#    geom_vline(xintercept = 127638255, linetype == "dashed") +
    scale_fill_manual(name = "State", values = meth_pal) + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), panel.spacing = unit(2, "lines"))
  
  print(g)

ggsave(
  paste0(figs, "/chr21_SST1_SingleRead.pdf"),
  plot = g,
  scale = 1,
  width = 10,
  height = 8,
  dpi = 50
)

```


```{r cen, echo=T, fig.height=5, fig.width=6}
# compare centromeric SST1 to non-centromeric SST1

# centromere boundaries
cen <- read_tsv(paste0(dat, "/annotations/t2t-chm13.v1.0.cenSat_regions.bed"), col_names = c("chr", "start", "end","name"))
cen.gr <- GRanges(cen)

chm13_meth <- readRDS(paste0(dat, "/methylation_calls/chm13_methylation_NanopolishFreq_50kb.rds"))
chm13_CpG <- readRDS(paste0(dat, "/reference/chm13_sliding_200_CpG.rds"))

#repeatmasker
rm <- read_tsv(paste0(dat, "/annotations/chm13.draft_v1.0.fasta_rm.bed"), col_names = F) %>%
  mutate(name = ifelse(X7 == "Satellite", X4, X7)) %>%
  filter(name == "SST1") %>%
  filter(X5 > 1290) %>%
  mutate(ID = row_number()) %>%
  dplyr::rename("chr"=1, "start"=2, "end"=3) %>%
  GRanges()

keepi <- findOverlaps(rm,cen.gr)
freq.matched <- rm[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
  mcols(freq.matched),
  mcols(cen.gr[subjectHits(keepi)]))

sst <- as.data.frame(rm) %>%
  mutate(region = case_when(ID %in% freq.matched$ID ~ "centr", 
                            TRUE ~ "non-centr")) %>%
  group_by(seqnames, region) %>%
  filter(n() > 10)

#write.table(sst, file = paste0(dat, "/TE/SST1_Cen_NonCen.bed"), append = FALSE, quote = F, sep = "\t", row.names = F, col.names = F)

sst.gr <- GRanges(sst)

keepi <- findOverlaps(chm13_meth,sst.gr)
freq.matched <- chm13_meth[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
  mcols(freq.matched),
  mcols(sst.gr[subjectHits(keepi)]))

sst_meth <- as.data.frame(freq.matched)
    


violin <- ggplot(data = sst_meth, aes(y = factor(seqnames), x = methylated_frequency, fill = region))+geom_density_ridges(rel_min_height = 0.01,scale = 0.9,quantile_lines = TRUE, quantiles = 2, bandwidth = .025)+theme_classic(base_size = 20)+labs(x = "Methylation", y = "Repeat")


ggsave(
  paste0(figs, "/SST1_CenvsNonCen.pdf"),
  plot = violin,
  scale = 1,
  width = 6,
  height = 5,
  dpi = 50
)


violin <- ggplot(data = sst_meth, aes(x = factor(region), y = methylated_frequency, fill = region))+geom_violin()+labs(x = "Methylation", y = "Repeat")+geom_boxplot(width=.1)+theme_classic()

ggsave(
  paste0(figs, "/SST1_CenvsNonCen_violin.pdf"),
  plot = violin,
  scale = 1,
  width = 6,
  height = 5,
  dpi = 50
)



keepi <- findOverlaps(chm13_CpG,sst.gr)
freq.matched <- chm13_CpG[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
  mcols(freq.matched),
  mcols(sst.gr[subjectHits(keepi)]))

sst_meth <- as.data.frame(freq.matched)

  

violin <- ggplot(data = sst_meth, aes(y = factor(region), x = CpG, fill = region))+geom_density_ridges(rel_min_height = 0.01,scale = 0.9,quantile_lines = TRUE, quantiles = 2, bandwidth = 1)+theme_classic(base_size = 20)+labs(x = "Methylation", y = "Repeat")

ggsave(
  paste0(figs, "/SST1_CenvsNonCen_CpG.pdf"),
  plot = violin,
  scale = 1,
  width = 6,
  height = 5,
  dpi = 50
)

violin <- ggplot(data = sst_meth, aes(x = factor(region), y = CpG, fill = region))+geom_violin()+labs(x = "CpG Density", y = "Repeat")+geom_boxplot(width=.1)+theme_classic()

ggsave(
  paste0(figs, "/SST1_CenvsNonCen_CpG_violin.pdf"),
  plot = violin,
  scale = 1,
  width = 6,
  height = 5,
  dpi = 50
)

```

