---
title: "TE quartiles"
author: "Ariel"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  pdf_document: default
  html_document:
    df_print: paged
  html_notebook: default
mainfont: DejaVu Sans
titlefont: DejaVu Sans
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE,  message=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(wesanderson)
library(ggExtra)
library(biovizBase)
library(Repitools)
library(Biostrings)
library(ggplot2)
library(BSgenome)
library(png)
library(cowplot)
options(scipen=999)
library(zoo)
library(ggridges)
library(GenomicRanges)
library(RColorBrewer)
library(BSgenome.t2t.v1.0.release)
options(knitr.duplicate.label = 'allow')
source("/home/isac/Code/ilee/plot/ilee_plot_utils.R")
library("ggsci")
library(gridExtra) 
library(pheatmap)
source("/home/isac/Code/nanopore-methylation-utilities/methylation_R_utils.R")
```

```{r loadData, include=F}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/TE/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"
chm13_meth <- readRDS(paste0(dat, "/methylation_calls/chm13_methylation_50kb.rds"))

L1quart <- data.frame()
for (i in 1:4){
  q <- read_tsv(paste0(dat, "/TE/transcription_quartiles/chm13-v1.0_PROseq-Meryl_L1_Transcribed_QUARTILE-", i, ".bed"), col_names = c("chr", "start", "end", "rep_type", "rep_class", "rep")) %>%
    mutate(width = end - start) %>%
    mutate(quart = as.factor(i))
  L1quart <- rbind(q, L1quart)
}

not <- read_tsv(paste0(dat, "/TE/transcription_quartiles/chm13-v1.0_PROseq-Meryl_L1_NotTranscribed.bed"), col_names = c("chr", "start", "end", "rep_type", "rep_class", "rep")) %>%
    mutate(width = end - start) %>%
    mutate(quart = as.factor("Not-transcribed"))


L1quart <- L1quart %>%
  rbind(L1quart, not) %>%
#  filter(width > 5500) %>%
  arrange(chr, start) %>%
  mutate(ID = row_number()) %>%
  select(c(chr, start, end, rep_type, width, ID, quart)) %>%
  na.omit() 

censat.gr <- read_tsv(paste0(dat, "/annotations/t2t-chm13.v1.0.cenSat_regions.bed"), col_names = c("chr", "start", "end", "name")) %>%
  GRanges()
```



```{r size, echo=F}
L1.gr <- GRanges(L1quart)
keepi <- findOverlaps(chm13_meth,L1.gr)
freq.matched <- chm13_meth[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
    mcols(freq.matched),
    mcols(L1.gr[subjectHits(keepi)]))

L1_meth <- as.data.frame(freq.matched) %>%
  group_by(ID) %>%
  mutate(avgmeth = mean(meth)) %>% 
  mutate(width = max(end)- min(start)) %>%
  select(width, avgmeth, rep_type, ID, quart) %>%
  distinct()

ggplot(L1_meth, aes(x = width, y = as.factor(quart), fill = as.factor(quart)))+geom_density_ridges()+xlim(0,6000)

ggsave(
  paste0(figs, "/L1_methylation_density"),
  plot = L1_meth,
  scale = 1,
  width = 5,
  height = 8,
)

```

```{r}
L1.gr <- L1quart %>%
  rename(width= "size") %>%
  GRanges()

keepi <- findOverlaps(chm13_meth,L1.gr)
freq.matched <- chm13_meth[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
    mcols(freq.matched),
    mcols(L1.gr[subjectHits(keepi)]))

L1_activity <- as.data.frame(freq.matched) %>%
  select(-c(cov)) %>%
  #filter(size > 5000) %>%
  distinct() 

violin <- ggplot(data = L1_activity, aes(y = factor(quart), x = meth, fill = quart))+geom_density_ridges(rel_min_height = 0.01,scale = 0.9,quantile_lines = TRUE, quantiles = 2, bandwidth = .01)+theme_classic(base_size = 20)+labs(x = "Methylation", y = "Repeat")

```



```{r SD_all, echo=F}

# SDs all 
rm <- read_tsv(paste0(dat, "/annotations/chm13.draft_v1.0.fasta_rm.bed"), col_names = F) %>%
  filter(X8 == "L1") %>%
  filter(X5 > 5500) %>%
  dplyr::rename("chr" = 1, "start"=2, "end"=3, "rep"=4, "width"=5, "direction"=6, "rep_type"=7, "rep_class"=8)

flankn <- 500
bodylen <- 2

l1_regions <- rm %>%
  mutate(start = start - flankn, end = end + flankn) %>%
  GRanges()

ovl <- findOverlaps(chm13_meth, l1_regions)
genes.ovl <- as.data.frame(rm)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl <- as.data.frame(chm13_meth[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  mutate(dist = ifelse(direction == "-",gene_end - start, start - gene_start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    )

chm13.ovl.labs <- chm13.ovl %>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
  group_by(dist) %>%
  summarise(med_meth = median(meth)) %>%
  distinct()

p <- ggplot(chm13.ovl.labs,aes( x = dist, y = med_meth)) +
  geom_smooth(method = "loess", span = 0.4,se = F)+
 # geom_smooth(se=F)+lims( y = c(0,1)) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen) +
  scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1),
    labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) +
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()#+facet_wrap(~seqnames, scales = "free")
p

#ggsave(
#  paste0(figs, "/","L1_quartiles_meta.pdf"),
#  plot = p,
#  scale = 1,
#  width = 10,
#  height = 10,
#)

```

```{r}
rm <- rm %>%
  mutate(ID = row_number())

L1.gr <- rm %>%
  mutate(start = start - flankn, end = end + flankn) %>%
  GRanges()

keepi <- findOverlaps(L1.gr,censat.gr)
matched <- L1.gr[queryHits(keepi)]

cen_ID <- as.data.frame(matched) %>%
  select(ID) %>%
  distinct()

age <-  read_tsv(paste0(dat, "/annotations/RM_L1_evage.bed"), col_names = c("rep", "age"))



ovl <- findOverlaps(chm13_meth, L1.gr)
genes.ovl <- as.data.frame(rm)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl <- as.data.frame(chm13_meth[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  mutate(dist = ifelse(direction == "-",gene_end - start, start - gene_start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    )

chm13.ovl.labs <- chm13.ovl %>%
  mutate(reg = ifelse(ID %in% cen_ID$ID, "cen", "non-cen")) %>%
  mutate(chr_type = case_when(seqnames == "chrX" ~ "chrX",
         TRUE ~ "autosome")) %>%
   mutate(lineage = case_when(grepl("L1M",rep) ~ "mouse",
         grepl("L1P",rep) ~ "primate", 
         TRUE ~ "other")) %>%
  group_by(dist, chr_type, reg) %>%
  summarise(med_meth = median(meth)) %>%
  distinct()

#L1_meth <- merge(chm13.ovl.labs, age, by = "rep")
# the 5' UTR of LINE elements is where the promoters are its the first 1kb of the the LINE (1/6) .17

p <- ggplot(chm13.ovl.labs,aes( x = dist, y = med_meth, color = reg)) +
  geom_smooth(method = "loess", span = 0.3,se = F)+
 # geom_smooth(se=F)+lims( y = c(0,1)) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen) +
  geom_vline(xintercept = bodylen/6) +
 # geom_rect(aes(xmin = 0, xmax = bodylen/6, ymin = 0, ymax = .4),fill = "grey", alpha = 0.01) + 
  scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1),
    labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) +
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+facet_grid(vars(chr_type), scales = "free")+ylim(.1,.4)
p
```



```{r}

L1.gr <- L1quart %>%
  filter(quart ==4 ) %>%
  GRanges()
keepi <- findOverlaps(chm13_meth,L1.gr)
freq.matched <- chm13_meth[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
    mcols(freq.matched),
    mcols(L1.gr[subjectHits(keepi)]))

L1_meth <- as.data.frame(freq.matched) %>%
  group_by(ID) %>%
  mutate(avgmeth = mean(meth)) %>% 
  mutate(end = max(end), start= min(start)) %>%
  select(seqnames, start,end, avgmeth, rep_type, ID, quart) %>%
  distinct() %>%
  dplyr::rename("chromosome" = seqnames)



chrom.sizes <- read_tsv(paste0(dat, "/reference/chrom.sizes"), col_names = F)
cen <- as.data.frame(censat.gr) %>%
  select(-c("strand", "name"))

colnames(chrom.sizes) <- c("chromosome", "size")
colnames(cen) <- c("chromosome", "start", "end", "width")

chrom_order <- c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", 
                 "chr8", "chr9", "chr10", "chr11", "chr12", "chr13", "chr14", 
                 "chr15", "chr16", "chr17", "chr18", "chr19", "chr20", "chr21", 
                 "chr22", "chrX")
chrom_key <- setNames(object = as.character(c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 
                                              12, 13, 14, 15, 16, 17, 18, 19, 20, 
                                              21, 22, 23)), 
                      nm = chrom_order)
chrom_order <- factor(x = chrom_order, levels = rev(chrom_order))

# convert the chromosome column in each dataset to the ordered factor
chrom.sizes[["chromosome"]] <- factor(x = chrom.sizes[["chromosome"]], 
                                      levels = chrom_order)
cen[["chromosome"]] <- factor(x = cen[["chromosome"]], 
                                      levels = chrom_order)

L1_meth[["chromosome"]] <- factor(x = L1_meth[["chromosome"]], 
                                     levels = chrom_order)

plot <- ggplot(data = chrom.sizes) + 
    # base rectangles for the chroms, with numeric value for each chrom on the x-axis
    geom_rect(aes(xmin = as.numeric(chromosome) - 0.2, 
                  xmax = as.numeric(chromosome) + 0.2, 
                  ymax = size, ymin = 0), 
              colour="black", fill = "white") + 
    # rotate the plot 90 degrees
    coord_flip() +
    # black & white color theme 
    theme(axis.text.x = element_text(colour = "black"), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank()) + 
    # give the appearance of a discrete axis with chrom labels
    scale_x_discrete(name = "chromosome", limits = names(chrom_key)) +
    # add bands for centromeres
    geom_rect(data = cen, aes(xmin = as.numeric(chromosome) - 0.2, 
                                      xmax = as.numeric(chromosome) + 0.2, 
                                      ymax = end, ymin = start, alpha = .5)) +
    # add bands for CNA value
    geom_rect(data = L1_meth, aes(xmin = as.numeric(chromosome) - 0.2, 
                                     xmax = as.numeric(chromosome) + 0.2, 
                                     ymax = end, ymin = start, fill = avgmeth)) + 
    scale_fill_gradient(low = "blue", high = "red") 

ggsave(
  paste0(figs, "/","L1_ideogram.pdf"),
  plot = plot,
  scale = 1,
  width = 10,
  height = 10,
)
```


