---
title: "TE quartiles"
author: "Ariel"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  pdf_document: default
  html_document:
    df_print: paged
  html_notebook: default
mainfont: DejaVu Sans
titlefont: DejaVu Sans
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE,  message=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(wesanderson)
library(ggExtra)
library(biovizBase)
library(Repitools)
library(Biostrings)
library(ggplot2)
library(BSgenome)
library(png)
library(cowplot)
options(scipen=999)
library(zoo)
library(ggridges)
library(GenomicRanges)
library(RColorBrewer)
library(BSgenome.t2t.v1.0.release)
options(knitr.duplicate.label = 'allow')
source("/home/isac/Code/ilee/plot/ilee_plot_utils.R")
library("ggsci")
library(gridExtra) 
library(pheatmap)
source("/home/isac/Code/nanopore-methylation-utilities/methylation_R_utils.R")
```

```{r loadData, include=F}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"
hg002_meth <- read_csv(paste0(dat, "/HG002/nanonome/methylation_calls/pooled/HG002_nanonome_chrX_CpGmethylation_ByCall.csv")) %>%
  ungroup() %>%
  group_by(start) %>%
  summarise(num_meth = sum(mcall == 1), num_unmeth = sum(mcall == 0)) %>%
  mutate(chr = "chrX", start = start, end = start) %>%
  GRanges()

hg002_gc <- read_csv(paste0(dat, "/HG002/nanonome/methylation_calls/pooled/HG002_nanonome_chrX_GpCmethylation_ByCall.csv")) %>%
  ungroup() %>%
  group_by(start) %>%
  summarise(num_meth = sum(mcall == 1), num_unmeth = sum(mcall == 0)) %>%
  mutate(chr = "chrX", start = start, end = start) %>%
  GRanges()
```

```{r proseq, echo=F}
r1 <- read_tsv(paste0(dat,"/HG002/annotations/hg002.chrX.fasta_rm.bed"), col_names = c("seqnames", "start", "end","rep_type", "len", "direction", "rep_family", "rep_class", "ID"))

rm <-  r1 %>%
  filter(rep_class == "Alu") %>%
  mutate(ID = row_number()) %>%
  group_by(rep_type) %>%
  ungroup()

flankn <- 1000
bodylen <- 2

l1_regions <- rm %>%
  mutate(start = start - flankn, end = end + flankn) %>%
  GRanges()

ovl <- findOverlaps(hg002_meth, l1_regions)
genes.ovl <- as.data.frame(rm)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl <- as.data.frame(hg002_meth[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  dplyr::mutate(dist = ifelse(direction == "+",start - gene_start, gene_end - start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    )

chm13.ovl.labs <- chm13.ovl %>%
  mutate(freq = num_meth/(num_unmeth+num_meth)) %>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
  group_by(dist) %>%
  summarise(med_meth = median(freq)) %>%
  mutate(mod = "CpG") %>%
  distinct()

p <- ggplot(chm13.ovl.labs,aes( x = dist, y = med_meth))+
  geom_smooth(method = "loess", span = 0.15,se = F)+
#  geom_smooth(se=T)+lims( y = c(0,1)) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen) +
  scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1),
    labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) +
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+ylim(0,1)
p

#ggsave(
#  paste0(figs, "/evol_meth/TE/","Alu_all_quartiles_meta_byname.pdf"),
#  plot = p,
#  scale = 1,
#  width = 10,
#  height = 10,
#)

```


```{r}
ovl <- findOverlaps(hg002_gc, l1_regions)
genes.ovl <- as.data.frame(rm)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl <- as.data.frame(hg002_gc[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  dplyr::mutate(dist = ifelse(direction == "+",start - gene_start, gene_end - start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    )


chm13.ovl.labsGC <- chm13.ovl %>%
  mutate(freq = (num_meth/(num_unmeth+num_meth)))%>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
  group_by(dist) %>%
  summarise(med_meth = median(freq)) %>%
  mutate(med_meth = med_meth*3.5) %>%
  mutate(mod = "GpC") %>%
  distinct()
  
ovl_all <- rbind(chm13.ovl.labs,chm13.ovl.labsGC)


GpC_unmeth <- chm13.ovl %>%
  group_by(start) %>%
  mutate(meth_freq = num_meth/(num_meth+num_unmeth)) %>%
  mutate(dist = ifelse(direction == "+",start - gene_start, gene_start - start)) %>%
  ungroup() %>%
  group_by(dist) %>%
  summarise(med_meth = median(meth_freq)) %>%
  distinct()
  
ggplot(GpC_unmeth, aes(x = dist, y = med_meth, alpha=.05))+ylim(0,1)+geom_smooth(method = "loess", span = 0.1,se = F)+xlim(-500,500)+coord_cartesian(ylim=c(.1,.3))+geom_vline(xintercept = 0, linetype = "dashed")+geom_point(alpha = .1)

```

