---
title: "Methylation and Divergence"
author: "Ariel"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  pdf_document: default
  html_document:
    df_print: paged
  html_notebook: default
mainfont: DejaVu Sans
titlefont: DejaVu Sans
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE,  message=FALSE)
library(knitr)
library(tidyverse)
library(ggplot2)
library(BSgenome)
options(scipen=999)
library(GenomicRanges)
```

```{r loadData, include=F}

figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"
chm13_meth <- read_tsv(paste0(dat, "/methylation_calls/methylation_frequency_50kb_split.tsv")) %>%
  GRanges()

regs <- read_delim(paste0(dat, "/annotations/chm13.draft_v1.0.fasta.out"), col_names = F, skip = 3, " ") %>%
  filter(X10 == "SST1") %>%
  rename( X5 = "seqnames", X6 = "start", X7 = "end" , X2 = "perc_div" ) %>%
  select(seqnames, start, end, perc_div) %>%
  mutate(ID = row_number()) %>%
  GRanges()

r1 <- read_tsv(paste0(dat, "/TE/AllChrs_IlluminaPCRfree_20200922-assembly_21mers_single_meryl.over.CHM13-5A_proseq_cutadapt-2_bt2-vs-filtered_bt2-N0-k100-chm13-20200922_COV_RMoutEx.bed"), col_names = c("chr", "start", "end", "len", "comp", "rep_type", "rep_fam", "col8", "col9", "col10", "col11", "SW", "perc_sub", "del", "ins", "cov", "num_depth", "len", "per_cov"))


censat = read_tsv(paste0(dat, "/annotations/t2t-chm13.v1.0.cenSat_annotation.bed"), col_names = F) %>%
  mutate(name = ifelse(grepl("HSat1", X4), "HSAT1", X4)) %>%
  mutate(name = ifelse(grepl("HSat2", X4), "HSAT2", name)) %>%
  mutate(name = ifelse(grepl("HSat3", X4), "HSAT3", name)) %>%
  mutate(name = ifelse(grepl("HSat4", X4), "HSAT4", name)) %>%
  mutate(name = ifelse(grepl("HSat5", X4), "HSAT5", name)) %>%
  mutate(name = ifelse(grepl("HSat6", X4), "HSAT6", name)) %>%
  mutate(name = ifelse(grepl("ct", X4), "CT", name)) %>%
  mutate(name = ifelse(grepl("bsat", X4), "BSAT", name)) %>%
  mutate(name = ifelse(grepl("hor", X4), "HOR", name)) %>%
  mutate(name = ifelse(grepl("mon", X4), "MON", name)) %>%
  mutate(name = ifelse(grepl("Alu", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("SATR", X4), "SATR", name)) %>%
  mutate(name = ifelse(grepl("ACRO", X4), "ACRO", name)) %>%
  mutate(name = ifelse(grepl("GSATII", X4), "GSATII", name)) %>%
  mutate(name = ifelse(grepl("TAR", X4), "TAR", name)) %>%
  mutate(name = ifelse(grepl("TE", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("MER", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("MST", X4), "MST", name)) %>%
  mutate(name = ifelse(grepl("CER", X4), "CER", name)) %>%
  mutate(name = ifelse(grepl("L1", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("SST", X4), "SST", name)) %>%
  mutate(name = ifelse(grepl("LSAU", X4), "LSAU", name)) %>%
  mutate(name = ifelse(grepl("GSAT", X4), "GSAT", name)) %>%
  mutate(name = ifelse(grepl("MSAT", X4), "MSAT", name)) %>%
  mutate(name = ifelse(grepl("novel", X4), "novel", name)) %>%
  mutate(name = ifelse(grepl("HERV", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("LT", X4), "TE", name)) %>%
  dplyr::select(c(X1, X2, X3, name)) %>%
  dplyr::rename("chrom" =1, "start" = 2 ,"end" =3) %>%
  group_by(name) %>%
  filter(n() >= 3) %>%
  ungroup() 

alpha.gr <- censat %>%
  filter(name %in% c("HOR", "MON")) %>%
  GRanges()
```

```{r alpha, echo=F}
alpha <- r1 %>%
  filter(rep_type == "ALR/Alpha") %>%
  GRanges()

keepi <- findOverlaps(chm13_meth,alpha.gr)

freq.matched <- chm13_meth[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
  mcols(freq.matched),
  mcols(alpha.gr[subjectHits(keepi)]))

alpha.meth <- as.data.frame(freq.matched) 

violin <- ggplot(data = alpha.meth, aes(x = as.factor(name), y = methylated_frequency, fill = as.factor(name)))+geom_violin()+labs(y = "Methylation Frequency")+geom_boxplot(width=.1, outlier.shape = NA)
violin

ggsave(
  paste0(figs, "/evol_meth/TE/","AlphaHORvsMON_methylation.pdf"),
  plot = violin,
  scale = 1,
  width = 10,
  height = 10,
)

```



```{r alpha, echo=F}
alpha <- r1 %>%
  filter(rep_type == "SST1") %>%
  GRanges()

keepi <- distanceToNearest(alpha)
freq.matched <- alpha[queryHits(keepi)]


mcols(freq.matched)  <- cbind.data.frame(
  mcols(freq.matched),
  mcols(keepi)) %>%
  mutate(type = case_when(distance > 1000 ~ "MON", 
                          TRUE ~ "Array")) 
array.gr <- freq.matched

keepi <- findOverlaps(chm13_meth,array.gr)
freq.matched <- chm13_meth[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
  mcols(freq.matched),
  mcols(array.gr[subjectHits(keepi)]))

sst_meth <- as.data.frame(freq.matched)


violin <- ggplot(data = sst_meth, aes(x = as.factor(type), y = methylated_frequency, fill = as.factor(type)))+geom_violin()+labs(y = "Methylation Frequency")+geom_boxplot(width=.1, outlier.shape = NA)
violin

ggsave(
  paste0(figs, "/evol_meth/TE/","SST1HORvsMON_methylation.pdf"),
  plot = violin,
  scale = 1,
  width = 10,
  height = 10,
)
```




