---
title: "HSAT4 HG002 vs chm13"
author: "G$"
date: "9/3/2020"
output: pdf_document
---

---
```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(Biostrings)
library(ggplot2)
library(png)
library(Repitools)
library(cowplot)
options(scipen=999)
library(zoo)
options(knitr.duplicate.label = 'allow')
source("~/projects/chm13_meth/scripts/v1.0_final_assembly/utils/ilee_plot_utils.R")
library("ggsci")
source("~/projects/chm13_meth/scripts/v1.0_final_assembly/utils/methylation_R_utils.R")
```

```{r mbed}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/censat/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"

censat = read_tsv(paste0(dat, "/annotations/t2t-chm13.v1.0.cenSat_annotation.bed"), col_names = F) %>%
  mutate(name = ifelse(grepl("HSat1", X4), "HSAT1", X4)) %>%
  mutate(name = ifelse(grepl("HSat2", X4), "HSAT2", name)) %>%
  mutate(name = ifelse(grepl("HSat3", X4), "HSAT3", name)) %>%
  mutate(name = ifelse(grepl("HSat4", X4), "HSAT4", name)) %>%
  mutate(name = ifelse(grepl("HSat5", X4), "HSAT5", name)) %>%
  mutate(name = ifelse(grepl("HSat6", X4), "HSAT6", name)) %>%
  mutate(name = ifelse(grepl("ct", X4), "CT", name)) %>%
  mutate(name = ifelse(grepl("bsat", X4), "BSAT", name)) %>%
  mutate(name = ifelse(grepl("hor", X4), "HOR", name)) %>%
  mutate(name = ifelse(grepl("mon", X4), "MON", name)) %>%
  mutate(name = ifelse(grepl("Alu", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("SATR", X4), "SATR", name)) %>%
  mutate(name = ifelse(grepl("ACRO", X4), "ACRO", name)) %>%
  mutate(name = ifelse(grepl("GSATII", X4), "GSATII", name)) %>%
  mutate(name = ifelse(grepl("TAR", X4), "TAR", name)) %>%
  mutate(name = ifelse(grepl("TE", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("MER", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("MST", X4), "MST", name)) %>%
  mutate(name = ifelse(grepl("CER", X4), "CER", name)) %>%
  mutate(name = ifelse(grepl("L1", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("SST", X4), "SST", name)) %>%
  mutate(name = ifelse(grepl("LSAU", X4), "LSAU", name)) %>%
  mutate(name = ifelse(grepl("GSAT", X4), "GSAT", name)) %>%
  mutate(name = ifelse(grepl("MSAT", X4), "MSAT", name)) %>%
  mutate(name = ifelse(grepl("novel", X4), "novel", name)) %>%
  mutate(name = ifelse(grepl("HERV", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("LT", X4), "TE", name)) %>%
  dplyr::select(c(X1, X2, X3, name)) %>%
  dplyr::rename("chrom" =1, "start" = 2 ,"end" =3) %>%
  group_by(name) %>%
  filter(n() >= 3) %>%
  ungroup()
table(censat$name)
censat.gr <- GRanges(as.data.frame(censat))
repeatColors =c("HSAT2"="#C19935",
                "HSAT1"="#A2A638",
                "HSAT3"="#8CAC3E",
                "(CATTC)n" = "#E87C71",
                "HSAT4"="#53B0E3",
                "HSAT5"="#ED72A5",
                "HSAT6"="#EF768C", 
                "(GAATC)n"="#E28455",
                "ALR/Alpha"="#D78C32",
                "6kbHsap" = "#D8BFD8",
                "BSR/Beta"="#E370AB",
                "CER" = "#CE9334",
                "DNA"="#C19935",
                "DNA?"="#C19935",
                "GSAT"="#4169E1",
                "LINE"="#FFA500",
                "Low_complexity"="#75B042",
                "LSAU"="#54B346",
                "LTR"="#51B756",
                "RC"="#53BB73",
                "Retroposon"="#55BE8D",
                "RNA"="#54C0A5",
                "rRNA"="#52BEBB",
                "SAR"="#51BDCE",
                "SATR2"="#4EB8DF",
                "GSATII"="#6B8E23",
                "SATR1"="#5AA5DA",
                "scRNA"="#6B9AD2",
                "Simple_repeat"="#8992C8",
                "SINE"="#9A8AC1",
                "snRNA"="#A885BC",
                "srpRNA"="#B67EB6",
                "SST1"="#C378B2",
                "HSATII"="#D173AF",
                "tRNA"="#006400",
                "ACRO1"="#9400D3",
                "Unknown"="#BA55D3", 
                "(GAATG)n"="#ff4000",
                "D20S16" = "#ffbf00", 
                "SATR2"= "#0080ff", 
                "TAR1" = "#000080", 
                "SUBTEL2_sat"= "#FFB6C1", 
                "GSATX" = "#D2691E", 
                "MSR1" = "#708090")
binnedSum <- function(bins, numvar, mcolname)
{
  stopifnot(is(bins, "GRanges"))
  stopifnot(is(numvar, "RleList"))
  stopifnot(identical(seqlevels(bins), names(numvar)))
  bins_per_chrom <- split(ranges(bins), seqnames(bins))
  sums_list <- lapply(names(numvar),
      function(seqname) {
          views <- Views(numvar[[seqname]],
                         bins_per_chrom[[seqname]])
          viewSums(views)
      })
  new_mcol <- unsplit(sums_list, as.factor(seqnames(bins)))
  mcols(bins)[[mcolname]] <- new_mcol
  bins
}

```


```{r plot, echo=F}
chr="chrX"

reads <- tabix_mbed(paste0(dat, "/censat/", chr, "_cen.mbed"),extcol = "motif",by = "read") 
#129004453       142241828 



size_sel <- reads %>%
  mutate(rlen = end-start) %>%
  filter(rlen >= 50000) %>%
  mutate(motif = "CG")

cgcalls <- mbedByCall(size_sel) %>%
  drop_na(mcall)

region_start=57379462
region_end= 57386639
reg = cgcalls %>%
  dplyr::filter(start >= region_start) %>%
  dplyr::filter(end <= region_end)
  
cpg_runs <-getRuns(reg, maxGap = 100)
 

cpg_runs.ordered <- order_reads(cpg_runs)
  
 cpg_runs <- cpg_runs.ordered$x %>%
    mutate(m = ifelse(values == 1, "Methylated","Unmethylated")) %>%
    mutate(mod = "CpG")

pal <- pal_npg("nrc")(10)
meth_pal <- c(pal[8],pal[4]) 
#ids <- order(sapply(seq_along(gcruns.list),function(i){length(unique(gcruns.list$qname))}))


g <- ggplot(cpg_runs,aes(xmin = start, xmax = end, ymin = ymin, ymax = ymax)) + geom_rect(data = cpg_runs.ordered$bounds, fill = "grey80") + 
    geom_rect(aes(fill = m))  +
#    geom_vline(xintercept = 127638255, linetype == "dashed") +
    scale_fill_manual(name = "State", values = meth_pal) + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), panel.spacing = unit(2, "lines"))
  
  print(g)

```



```{r chmplot, echo=F}
region_start=57279462
region_end= 57654110

rm <- read_tsv(paste0(dat, "/annotations/chm13.draft_v1.0.fasta_rm.bed"), col_names = F) %>%
  mutate(name = ifelse(X7 == "Satellite", X4, X7))

library(BSgenome.t2t.v1.0.release)
cgcalls.gr <-  cgcalls %>%
  ungroup() %>%
  group_by(start) %>%
  summarise(num_meth = sum(mcall == 1), num_unmeth = sum(mcall == 0)) %>%
  mutate(chr = "chrX", start = start, end = start) %>%
  GRanges()

blocks <- genomeBlocks(BSgenome.t2t.v1.0.release, chrs = "chrX", width = 1000)
score1 <- coverage(cgcalls.gr, weight="num_meth")
score2 <- coverage(cgcalls.gr, weight="num_unmeth")

binned_meth <- binnedSum(blocks, numvar = score1, "num_meth") %>%
  as.data.frame()
binned_unmeth <-binnedSum(blocks, numvar = score2, "num_unmeth")%>%
  as.data.frame()

meth_bins <- merge(binned_meth, binned_unmeth, by = c("start", "end", "seqnames", "width", "strand")) %>%
  filter(start > region_start) %>%
  filter(end < region_end) %>%
  group_by(start, end) %>%
  mutate(sites = num_meth+num_unmeth) %>%
  mutate(freq = num_meth/sites) %>%
  na.omit()


meth_CG <- ggplot(meth_bins, aes(x = (start/1e6), y= freq)) + labs(x="Genomic coordinates (Mb)", y="Methylation")+theme_classic(base_size = 25)+ylim(0,1)+geom_line(size =1)


censat_reg <- rm %>%
  filter(X1 == "chrX") 

rep_leg <- ggplot(data=censat_reg, mapping=aes(xmin=(X2),xmax=(X3),ymin=0,ymax=.1, fill = name))+
  geom_rect()  +  theme_void() +theme(legend.position="top") +labs(y="Axis")+labs(y="Axis")+ scale_fill_manual(values = repeatColors, drop = FALSE)+theme(legend.text=element_text(size=rel(1.5)))+theme(legend.title=element_blank())+coord_cartesian(xlim= c(region_start, region_end))

top_row <- plot_grid(rep_leg,meth_CG, ncol = 1, align="v")
top_row
```



```{r bis, echo=F}


reads <- tabix_mbed(paste0(dat, "/HG002/nanonome/methylation_calls/HG002_nanonome_chrX_CpGmethylation.merge1-3_HSAT4.bed"),extcol = "motif",by = "read") 
#129004453       142241828 



region_start=55500894
region_end=55508071


size_sel <- reads %>%
  mutate(rlen = end-start) %>%
  filter(rlen >= 20000) %>%
  mutate(motif = "CG")

cgcalls <- mbedByCall(size_sel) %>%
  drop_na(mcall)


reg = cgcalls %>%
  dplyr::filter(start >= region_start) %>%
  dplyr::filter(end <= region_end)
  

cpg_runs <-getRuns(reg, maxGap = 100)
 

cpg_runs.ordered <- order_reads(cpg_runs)
  
 cpg_runs <- cpg_runs.ordered$x %>%
    mutate(m = ifelse(values == 1, "Methylated","Unmethylated")) %>%
    mutate(mod = "CpG")

pal <- pal_npg("nrc")(10)
meth_pal <- c(pal[8],pal[4]) 
#ids <- order(sapply(seq_along(gcruns.list),function(i){length(unique(gcruns.list$qname))}))


g <- ggplot(cpg_runs,aes(xmin = start, xmax = end, ymin = ymin, ymax = ymax)) + geom_rect(data = cpg_runs.ordered$bounds, fill = "grey80") + 
    geom_rect(aes(fill = m))  +
#    geom_vline(xintercept = 127638255, linetype == "dashed") +
    scale_fill_manual(name = "State", values = meth_pal) + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), panel.spacing = unit(2, "lines"))
  
  print(g)



```

```{r 002plot, echo=F}

region_start=55400894
region_end=55775544

reps <- read_tsv("/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/HG002/annotations/hg002.chrX.fasta_rm.bed", col_names = F) %>%
  mutate(name = ifelse(X7 == "Satellite", X4, X7))

library(BSgenome.HG002.chrX)
cgcalls.gr <-  cgcalls %>%
  ungroup() %>%
  group_by(start) %>%
  summarise(num_meth = sum(mcall == 1), num_unmeth = sum(mcall == 0)) %>%
  mutate(chr = "chrX", start = start, end = start) %>%
  GRanges()

blocks <- genomeBlocks(BSgenome.HG002.chrX, chrs = "chrX", width = 1000)
score1 <- coverage(cgcalls.gr, weight="num_meth")
score2 <- coverage(cgcalls.gr, weight="num_unmeth")

binned_meth <- binnedSum(blocks, numvar = score1, "num_meth") %>%
  as.data.frame()
binned_unmeth <-binnedSum(blocks, numvar = score2, "num_unmeth")%>%
  as.data.frame()

meth_bins <- merge(binned_meth, binned_unmeth, by = c("start", "end", "seqnames", "width", "strand")) %>%
  filter(start > region_start) %>%
  filter(end < region_end) %>%
  group_by(start, end) %>%
  mutate(sites = num_meth+num_unmeth) %>%
  mutate(freq = num_meth/sites) 
  


meth_CG <- ggplot(meth_bins, aes(x = (start/1e6), y= freq)) + labs(x="Genomic coordinates (Mb)", y="Methylation")+theme_classic(base_size = 25)+ylim(0,1)+geom_line(size =1)+xlim(region_start/1e6, region_end/1e6)



rep_leg <- ggplot(data=reps, mapping=aes(xmin=(X2),xmax=(X3),ymin=0,ymax=.1, fill = name))+
  geom_rect()  +  theme_void() +theme(legend.position="top") +labs(y="Axis")+labs(y="Axis")+ scale_fill_manual(values = repeatColors, drop = FALSE)+theme(legend.text=element_text(size=rel(1.5)))+theme(legend.title=element_blank())+coord_cartesian(xlim= c(region_start, region_end))

top_row <- plot_grid(rep_leg,meth_CG, ncol = 1, align="v")
top_row
```

