---
title: "Read plots"
author: "G$"
date: "9/3/2020"
output: pdf_document
---

---
```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(Biostrings)
library(ggplot2)
library(png)
library(cowplot)
options(scipen=999)
library(zoo)
options(knitr.duplicate.label = 'allow')
source("/home/isac/Code/ilee/plot/ilee_plot_utils.R")
library("ggsci")
source("/home/isac/Code/nanopore-methylation-utilities/methylation_R_utils.R")
```

```{r mbed}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/censat/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"

censat = read_tsv(paste0(dat, "/annotations/t2t-chm13.v1.0.cenSat_annotation.bed"), col_names = F) %>%
  mutate(name = ifelse(grepl("HSat1", X4), "HSAT1", X4)) %>%
  mutate(name = ifelse(grepl("HSat2", X4), "HSAT2", name)) %>%
  mutate(name = ifelse(grepl("HSat3", X4), "HSAT3", name)) %>%
  mutate(name = ifelse(grepl("HSat4", X4), "HSAT4", name)) %>%
  mutate(name = ifelse(grepl("HSat5", X4), "HSAT5", name)) %>%
  mutate(name = ifelse(grepl("HSat6", X4), "HSAT6", name)) %>%
  mutate(name = ifelse(grepl("ct", X4), "CT", name)) %>%
  mutate(name = ifelse(grepl("bsat", X4), "BSAT", name)) %>%
  mutate(name = ifelse(grepl("hor", X4), "HOR", name)) %>%
  mutate(name = ifelse(grepl("mon", X4), "MON", name)) %>%
  mutate(name = ifelse(grepl("Alu", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("SATR", X4), "SATR", name)) %>%
  mutate(name = ifelse(grepl("ACRO", X4), "ACRO", name)) %>%
  mutate(name = ifelse(grepl("GSATII", X4), "GSATII", name)) %>%
  mutate(name = ifelse(grepl("TAR", X4), "TAR", name)) %>%
  mutate(name = ifelse(grepl("TE", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("MER", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("MST", X4), "MST", name)) %>%
  mutate(name = ifelse(grepl("CER", X4), "CER", name)) %>%
  mutate(name = ifelse(grepl("L1", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("SST", X4), "SST", name)) %>%
  mutate(name = ifelse(grepl("LSAU", X4), "LSAU", name)) %>%
  mutate(name = ifelse(grepl("GSAT", X4), "GSAT", name)) %>%
  mutate(name = ifelse(grepl("MSAT", X4), "MSAT", name)) %>%
  mutate(name = ifelse(grepl("novel", X4), "novel", name)) %>%
  mutate(name = ifelse(grepl("HERV", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("LT", X4), "TE", name)) %>%
  dplyr::select(c(X1, X2, X3, name)) %>%
  dplyr::rename("chrom" =1, "start" = 2 ,"end" =3) %>%
  group_by(name) %>%
  filter(n() >= 3) %>%
  ungroup()
table(censat$name)
censat.gr <- GRanges(as.data.frame(censat))

repeatColors =c("(CATTC)n" = "#E87C71",
    "(GAATC)n"="#E28455",
    "HOR"="#D78C32",
    "BSAT"="#E370AB",
    "CER" = "#CE9334",
    "HSAT2"="#C19935",
    "HSAT1"="#A2A638",
    "HSAT3"="#8CAC3E",
    "Low_complexity"="#75B042",
    "LSAU"="#54B346",
    "LTR"="#51B756",
    "MST"="#53BB73",
    "GSAT"="#55BE8D",
    "RNA"="#54C0A5",
    "rRNA"="#52BEBB",
    "SAR"="#51BDCE",
    "ACRO"="#4EB8DF",
    "HSAT4"="#53B0E3",
    "SATR"="#5AA5DA",
    "CT"="#6B9AD2",
    "Simple_repeat"="#8992C8",
    "SINE"="#9A8AC1",
    "MON"="#A885BC",
    "SST"="#C378B2",
    "HSAT5"="#ED72A5",
    "HSAT6"="#EF768C", 
    "gap-rDNA"="#ff4000",
    "TE" = "#ffbf00", 
   "TAR"= "#0080ff" )

```


```{r plot, echo=F}
chr="chr1"

reads <- tabix_mbed(paste0(dat, "/censat/", chr, "_cen.mbed"),extcol = "motif",by = "read") 
#129004453       142241828 



size_sel <- reads %>%
  mutate(rlen = end-start) %>%
  filter(rlen >= 50000) %>%
  mutate(motif = "CG")

cgcalls <- mbedByCall(size_sel) %>%
  drop_na(mcall)

region_start=129604453
region_end= 129651828
reg = cgcalls %>%
  dplyr::filter(start >= region_start) %>%
  dplyr::filter(end <= region_end)
  

cpg_runs <-getRuns(reg, maxGap = 500)
 

cpg_runs.ordered <- order_reads(cpg_runs)
  
 cpg_runs <- cpg_runs.ordered$x %>%
    mutate(m = ifelse(values == 1, "Methylated","Unmethylated")) %>%
    mutate(mod = "CpG")

pal <- pal_npg("nrc")(10)
meth_pal <- c(pal[8],pal[4]) 
#ids <- order(sapply(seq_along(gcruns.list),function(i){length(unique(gcruns.list$qname))}))


bis <- reg %>% 
  group_by(chrom, start) %>%
  summarise(meth_freq = mean(mcall), cov = n()) 

meth <- ggplot(bis, aes(x = (start/1e6), y= meth_freq))+geom_smooth(size =1, se = F, method = "loess", span = .1)+ylim(0,1)+ theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.y=element_blank(), axis.title.x=element_blank(), panel.spacing = unit(2, "lines"))#+coord_cartesian(xlim=(c(37595000,37600000)))


g <- ggplot(cpg_runs,aes(xmin = start, xmax = end, ymin = ymin, ymax = ymax)) + geom_rect(data = cpg_runs.ordered$bounds, fill = "grey80") + 
    geom_rect(aes(fill = m))  +
#    geom_vline(xintercept = 127638255, linetype == "dashed") +
    scale_fill_manual(name = "State", values = meth_pal) + theme(axis.text.y = element_blank(),  legend.position= "bottom",axis.ticks.y = element_blank(), panel.spacing = unit(2, "lines"))

top_row <- plot_grid(meth,g, align="v", ncol=1,rel_heights = c(1/4, 1/2))
top_row
  
  ggsave(
  paste0(figs, "/", chr, "HSAT2_singleread.pdf"),
  plot = g,
  scale = 1,
  width = 10,
  height = 6,
)

```



```{r hsat1, echo=F}

chr="chr3"

reads <- tabix_mbed(paste0(dat, "/censat/", chr, "_cen.mbed"),extcol = "motif",by = "read") 
#129004453       142241828 



size_sel <- reads %>%
  mutate(rlen = end-start) %>%
  filter(rlen >= 50000) %>%
  mutate(motif = "CG")

cgcalls <- mbedByCall(size_sel) %>%
  drop_na(mcall)

region_start=93900000
region_end= 93956794

#chr3    92903598        95366794
reg = cgcalls %>%
  dplyr::filter(start >= region_start) %>%
  dplyr::filter(end <= region_end)
  

cpg_runs <-getRuns(reg, maxGap = 500)
 

cpg_runs.ordered <- order_reads(cpg_runs)
  
 cpg_runs <- cpg_runs.ordered$x %>%
    mutate(m = ifelse(values == 1, "Methylated","Unmethylated")) %>%
    mutate(mod = "CpG")

pal <- pal_npg("nrc")(10)
meth_pal <- c(pal[8],pal[4]) 
#ids <- order(sapply(seq_along(gcruns.list),function(i){length(unique(gcruns.list$qname))}))


bis <- reg %>% 
  group_by(chrom, start) %>%
  summarise(meth_freq = mean(mcall), cov = n()) 

meth <- ggplot(bis, aes(x = (start/1e6), y= meth_freq))+geom_smooth(size =1, se = F, method = "loess", span = .1)+ylim(0,1)+ theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.y=element_blank(), axis.title.x=element_blank(), panel.spacing = unit(2, "lines"))#+coord_cartesian(xlim=(c(37595000,37600000)))


g <- ggplot(cpg_runs,aes(xmin = start, xmax = end, ymin = ymin, ymax = ymax)) + geom_rect(data = cpg_runs.ordered$bounds, fill = "grey80") + 
    geom_rect(aes(fill = m))  +
#    geom_vline(xintercept = 127638255, linetype == "dashed") +
    scale_fill_manual(name = "State", values = meth_pal) + theme(axis.text.y = element_blank(),  legend.position= "bottom",axis.ticks.y = element_blank(), panel.spacing = unit(2, "lines"))

top_row <- plot_grid(meth,g, align="v", ncol=1,rel_heights = c(1/4, 1/2))
top_row
  
  
  ggsave(
  paste0(figs, "/", chr, "HSAT1_singleread.pdf"),
  plot = top_row,
  scale = 1,
  width = 10,
  height = 4,
)

```

```{r}

chm13_meth <- read_tsv(paste0(dat, "/methylation_calls/methylation_frequency_50kb_split.tsv")) %>%
  GRanges()

#repeatmasker
censat.gr <- GRanges(censat)

keepi <- findOverlaps(chm13_meth,censat.gr)
freq.matched <- chm13_meth[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
  mcols(freq.matched),
  mcols(censat.gr[subjectHits(keepi)]))

list = c("HSAT1", "HSAT2")
# violing plot of methylation frequency
rep_meth <- as.data.frame(freq.matched) %>%
  filter(name %in% list)

violin <- ggplot(data = rep_meth, aes(x = factor(seqnames), y = methylated_frequency, fill = name))+geom_violin()+facet_wrap(~name, scales = "free")+geom_boxplot(width=.1, outlier.shape = NA)+theme_classic()


#+geom_density_ridges(rel_min_height = 0.01,scale = 0.9,quantile_lines = TRUE, quantiles = 2, bandwidth = .015)+theme_classic(base_size = 20)+ scale_fill_manual(values = repeatColors, drop = FALSE)+labs(x = "Methylation", y = "Repeat")+facet_wrap(~name)

ggsave(
  paste0(figs, "/HSAT1_HSAT2_all_chrs.pdf"),
  plot = violin,
  scale = 1,
  width = 8,
  height = 5,
)

```
```{r}
# look at cen19 without filtering for read length to show methylation profile is very different
chr="chr19"

reads <- tabix_mbed(paste0(dat, "/censat/", chr, "_cen.mbed"),extcol = "motif",by = "read") 
#129004453       142241828 



size_sel <- reads %>%
  mutate(rlen = end-start) %>%
 # filter(rlen >= 50000) %>%
  mutate(motif = "CG")

cgcalls <- mbedByCall(size_sel) %>%
  drop_na(mcall)
```

