---
title: "Isoseq expression quartiles"
author: "Ariel"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  pdf_document: default
  html_document:
    df_print: paged
  html_notebook: default
mainfont: DejaVu Sans
titlefont: DejaVu Sans
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE,  message=FALSE)
library(knitr)
library(tidyverse)
library(ggplot2)
library(BSgenome)
options(scipen=999)
library(GenomicRanges)
```

```{r loadData, include=F}
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"
chm13_meth <- read_tsv(paste0(dat, "/methylation_calls/methylation_frequency_50kb_split.tsv")) %>%
  GRanges()

regs <- read_tsv(paste0(dat, "/sd/sd.transcripts.and.meth.bed")) %>%
  rename(`#chr` = "chr") %>%
  mutate(quartile = ntile(cov,4)) 
```

```{r 5kb, echo=F}


flankn <- 10000
bodylen <- 2

l1_regions <- regs %>%
  mutate(len=end-start) %>%
  mutate(start = start - flankn, end = end + flankn) %>%
  select(c(chr, start, end)) %>%
  GRanges()

ovl <- findOverlaps(chm13_meth, l1_regions)
genes.ovl <- as.data.frame(regs)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl <- as.data.frame(chm13_meth[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  mutate(dist = ifelse(strand...15 == "-", gene_end - start, start - gene_start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    )

df <- chm13.ovl %>%
  select(-c(strand...5)) %>%
  rename(strand...15 = "strand")
  
write.table(df, file =paste0(dat, "/sd/sd.transcripts.10kb_methAG.bed"),  quote = F, sep = "\t",row.names = F,
            col.names = T)


chm13.ovl.labs <- chm13.ovl %>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
  group_by(dist, quartile,is_sd) %>%
  summarise(med_meth = median(methylated_frequency)) %>%
  distinct()

p <- ggplot(chm13.ovl.labs,aes( x = dist, y = med_meth, color = as.factor(is_sd)))+
  geom_smooth(method = "loess", span = 0.3,se = F)+
#  geom_smooth(se=T)+lims( y = c(0,1)) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen) +
  scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1),
    labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) +
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+ylim(0,1)+facet_wrap(~quartile)
p


chm13.ovl.labs <- chm13.ovl %>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
  group_by(dist,is_sd) %>%
  summarise(med_meth = median(methylated_frequency)) %>%
  distinct()


write.table(chm13.ovl.labs, file =paste0(dat, "/sd/sd.transcripts.10kb_methAGForPlot.bed"),  quote = F, sep = "\t",row.names = F,
            col.names = T)

p <- ggplot(chm13.ovl.labs,aes( x = dist, y = med_meth, color = as.factor(is_sd)))+
  geom_smooth(method = "loess", span = 0.3,se = F)+
#  geom_smooth(se=T)+lims( y = c(0,1)) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen) +
  scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1),
    labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) +
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+ylim(0,1)
p


```










