---
title: "TE quartiles"
author: "Ariel"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  pdf_document: default
  html_document:
    df_print: paged
  html_notebook: default
mainfont: DejaVu Sans
titlefont: DejaVu Sans
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE,  message=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(wesanderson)
library(ggExtra)
library(biovizBase)
library(Repitools)
library(Biostrings)
library(ggplot2)
library(BSgenome)
library(png)
library(cowplot)
options(scipen=999)
library(zoo)
library(ggridges)
library(GenomicRanges)
library(RColorBrewer)
library(BSgenome.t2t.v1.0.release)
options(knitr.duplicate.label = 'allow')
source("/home/isac/Code/ilee/plot/ilee_plot_utils.R")
library("ggsci")
library(gridExtra) 
library(pheatmap)
source("/home/isac/Code/nanopore-methylation-utilities/methylation_R_utils.R")
```

```{r loadData, include=F}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"
hg002_meth <- read_csv(paste0(dat, "/HG002/nanonome/methylation_calls/pooled/HG002_nanonome_chrX_CpGmethylation_ByCall.csv")) %>%
  ungroup() %>%
  group_by(start) %>%
  summarise(num_meth = sum(mcall == 1), num_unmeth = sum(mcall == 0)) %>%
  mutate(chr = "chrX", start = start, end = start) %>%
  GRanges()

hg002_gc <- read_csv(paste0(dat, "/HG002/nanonome/methylation_calls/pooled/HG002_nanonome_chrX_GpCmethylation_ByCall.csv")) %>%
  ungroup() %>%
  group_by(start) %>%
  summarise(num_meth = sum(mcall == 1), num_unmeth = sum(mcall == 0)) %>%
  mutate(chr = "chrX", start = start, end = start) %>%
  GRanges()
r1 <- read_tsv(paste0(dat, "/TE/rm_v2.0/hg002_k51_chrX_single-mrg_meryl.over.HG002_RNAseq_SRR13086640_cutadapt-m100_bt2-N0-k100-HG002-chrX_COV_RM.bed"), col_names = c("chr", "start", "end", "len", "comp", "rep_type", "rep_fam", "col8", "col9", "col10", "col11", "SW", "perc_sub", "del", "ins", "cov", "num_depth", "len", "per_cov")) 

```



```{r}

list=c("AluJb", "AluSx", "AluY")
rm <- r1 %>%
  filter(rep_fam == "SINE/Alu") %>%
  filter(!(abs(perc_sub - median(perc_sub)) > 2*sd(perc_sub))) %>%
#  filter(rep_class == "Alu") %>%
  mutate(name = ifelse(grepl("AluJb", rep_fam), "AluJb", rep_fam)) %>%
  mutate(name = ifelse(grepl("AluSx", rep_fam), "AluSx", name)) %>%
  mutate(name = ifelse(grepl("AluY", rep_fam), "AluY", name)) %>%
  #filter(name != "SINE") %>%
  filter(rep_type %in% list) %>%
  group_by(perc_sub,rep_type) %>%
  mutate(status=case_when(cov == 0 ~ "untranscribed", 
                          TRUE ~ "transcribed")) %>%
  group_by(perc_sub,rep_type) %>%
mutate(quartile = as.factor(ntile(perc_sub, 3))) 


#L1.gr <- GRanges(L1quart)
#keepi <- findOverlaps(chm13_meth,L1.gr)
#freq.matched <- chm13_meth[queryHits(keepi)]
#
#mcols(freq.matched) <- cbind.data.frame(
#    mcols(freq.matched),
#    mcols(L1.gr[subjectHits(keepi)]))

flankn <- 1000

bodylen <- 300/1000
l1_regions <- rm %>%
  mutate(start = start - flankn, end = end + flankn) %>%
  GRanges()

ovl <- findOverlaps(hg002_meth, l1_regions)
genes.ovl <- as.data.frame(rm)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl <- as.data.frame(hg002_meth[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  dplyr::mutate(dist = ifelse(comp == "+",start - gene_start, gene_end - start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    )

n_windows=100
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
 # filter(quartile != 0) %>%
  group_by(cut,quartile,rep_type) %>%
  mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency)) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  group_by(rep_type,quartile) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 10, NA))


# 4 is lowest divergence (youngest) 1 is highest divergence (oldest)
p <- ggplot(chm13.ovl.labs,aes( x = min, y = med_smooth, color=quartile), alpha=.5)+
#  geom_smooth(method = "loess", span = 0.12,se = F)+
#  geom_smooth(se=T)+lims( y = c(0,1)) +
 # geom_ribbon(aes(ymin=bot, ymax=top), alpha=0.2, size=0)+
  geom_line(aes(y=med_smooth), alpha=.5, size=1) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+facet_wrap(~rep_type)+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb")))+ylim(.5,1)

#+
#geom_text(data = chm13.ovl %>% 
#              ungroup() %>%
#              select(ID,rep_type) %>%
#              distinct() %>%
#              group_by(rep_type) %>%
#              summarise(n_genes=n()) %>%
#              distinct() %>%
#              mutate(height=case_when(rep_type == "AluY" ~ 0, 
#                          rep_type == "AluJb" ~ .01, 
#                          TRUE ~ .02)),
#            aes(x=c(2),
#                y=.2+height,
#                label=paste("# repeats =",comma(n_genes))
#                ), hjust=1)+facet_wrap(~rep_type)
p

ggsave(
  paste0(figs, "/evol_meth/TE/","Alu_meta_age_HG002.pdf"),
  plot = p,
  scale = 1,
  width = 10,
  height = 5,
)
#scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1),
#    labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) 
```


```{r}
# histogram of fraction of elements active at each age

frac.active <- rm %>%
  #filter(!(abs(perc_sub - median(perc_sub)) > 2*sd(perc_sub))) %>%
  group_by(perc_sub) %>%
  summarise(frac = sum(status == "transcribed")/n())

hist <- ggplot(frac.active, aes(x=perc_sub, y=frac))+geom_histogram(stat="identity", position = "dodge")+coord_flip()+xlim(0,21)
```

```{r}


flankn <- 1000

n_windows=100
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
  #filter(!(abs(perc_sub - median(perc_sub)) > 2*sd(perc_sub))) %>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
 # filter(quartile != 0) %>%
  group_by(cut,perc_sub) %>%
  mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency)) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  group_by(perc_sub) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 4, NA))


plot <- ggplot(chm13.ovl.labs,aes(x=min,y=perc_sub,fill=med_smooth))+ geom_tile()+scale_fill_gradientn(colours = rev(rainbow(5)))+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) +geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+ylim(0,21)

plot_grid( plot,hist,ncol=2)

ggsave(
  paste0(figs, "/evol_meth/TE/","Alu_methAge_heatmapSmoothHG002.pdf"),
  plot = last_plot(),
  scale = 1,
  width = 10,
  height = 10,
)
```

```{r}

flankn <- 1000

bodylen <- 300/1000
l1_regions <- rm %>%
  mutate(start = start - flankn, end = end + flankn) %>%
  GRanges()

ovl <- findOverlaps(hg002_gc, l1_regions)
genes.ovl <- as.data.frame(rm)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl <- as.data.frame(hg002_gc[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  dplyr::mutate(dist = ifelse(comp == "+",start - gene_start, gene_end - start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    )

n_windows=1050
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
 # filter(quartile != 0) %>%
  group_by(cut) %>%
  mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency)) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
 # group_by(quartile) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 30, NA))


# 4 is lowest divergence (youngest) 1 is highest divergence (oldest)
p <- ggplot(chm13.ovl.labs,aes( x = min, y = med_smooth), alpha=.5)+
#  geom_smooth(method = "loess", span = 0.12,se = F)+
#  geom_smooth(se=T)+lims( y = c(0,1)) +
 # geom_ribbon(aes(ymin=bot, ymax=top), alpha=0.2, size=0)+
  geom_line(aes(y=med_smooth), alpha=.5, size=1) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
 theme(legend.background = element_rect(color = "black")) +
  theme_classic()+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb")))+ylim(.15,.25)
#+
#geom_text(data = chm13.ovl %>% 
#              ungroup() %>%
#              select(ID,quartile) %>%
#              distinct() %>%
#              group_by(rep_type) %>%
#              summarise(n_genes=n()) %>%
#             distinct() %>%
#              mutate(height=case_when(rep_type == "AluY" ~ 0, 
#                          rep_type == "AluJb" ~ .01, 
#                          TRUE ~ .02)),
#            aes(x=c(2),
#                y=.2+height,
#                label=paste("# repeats =",comma(n_genes))
#                ), hjust=1)+facet_wrap(~rep_type)
p

ggsave(
  paste0(figs, "/evol_meth/TE/","Alu_meta_age_HG002GC_quartiles_noname.pdf"),
  plot = p,
  scale = 1,
  width = 10,
  height = 5,
)
#scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1),
#    labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) 
```


```{r}


flankn <- 1000
bodylen <- 300/1000
n_windows=500
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
  #filter(!(abs(perc_sub - median(perc_sub)) > 2*sd(perc_sub))) %>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
 # filter(quartile != 0) %>%
  group_by(cut,perc_sub) %>%
  mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency)) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  group_by(perc_sub) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 20, NA))


plot <- ggplot(chm13.ovl.labs,aes(x=min,y=perc_sub,fill=med_smooth))+ geom_tile()+scale_fill_gradientn(colours = rev(rainbow(5)))+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) +geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+ylim(0,21)

plot_grid( plot,hist,ncol=2)

ggsave(
  paste0(figs, "/evol_meth/TE/","Alu_methAge_heatmapSmoothHG002GC_nucs.pdf"),
  plot = last_plot(),
  scale = 1,
  width = 10,
  height = 10,
)
```



```{r}
rm <- r1 %>%
  filter(!(abs(perc_sub - median(perc_sub)) > 2*sd(perc_sub))) %>%
  filter(rep_fam == "LINE/L1") %>%
  mutate(ID = row_number()) %>%
  #sample_n(500000) %>%
  mutate(status=case_when(cov == 0 ~ "untranscribed", 
                          TRUE ~ "transcribed")) %>%
  group_by(perc_sub) %>%
mutate(quartile = as.factor(ntile(perc_sub, 3))) 

#L1.gr <- GRanges(L1quart)
#keepi <- findOverlaps(chm13_meth,L1.gr)
#freq.matched <- chm13_meth[queryHits(keepi)]
#
#mcols(freq.matched) <- cbind.data.frame(
#    mcols(freq.matched),
#    mcols(L1.gr[subjectHits(keepi)]))

flankn <- 5000

bodylen <- 1
l1_regions <- rm %>%
  mutate(start = start - flankn, end = end + flankn) %>%
  GRanges()

ovl <- findOverlaps(hg002_meth, l1_regions)
genes.ovl <- as.data.frame(rm)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl <- as.data.frame(hg002_meth[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  dplyr::mutate(dist = ifelse(comp == "+",start - gene_start, gene_end - start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    )

n_windows=100
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
 # filter(quartile != 0) %>%
  group_by(cut,quartile) %>%
  mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency)) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  group_by(quartile) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 10, NA))


# 4 is lowest divergence (youngest) 1 is highest divergence (oldest)
p <- ggplot(chm13.ovl.labs,aes( x = min, y = med_smooth, color=quartile), alpha=.5)+
#  geom_smooth(method = "loess", span = 0.12,se = F)+
#  geom_smooth(se=T)+lims( y = c(0,1)) +
 # geom_ribbon(aes(ymin=bot, ymax=top), alpha=0.2, size=0)+
  geom_line(aes(y=med_smooth), alpha=.5, size=1) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb")))

#+
#geom_text(data = chm13.ovl %>% 
#              ungroup() %>%
#              select(ID,rep_type) %>%
#              distinct() %>%
#              group_by(rep_type) %>%
#              summarise(n_genes=n()) %>%
#              distinct() %>%
#              mutate(height=case_when(rep_type == "AluY" ~ 0, 
#                          rep_type == "AluJb" ~ .01, 
#                          TRUE ~ .02)),
#            aes(x=c(2),
#                y=.2+height,
#                label=paste("# repeats =",comma(n_genes))
#                ), hjust=1)+facet_wrap(~rep_type)
p

ggsave(
  paste0(figs, "/evol_meth/TE/","L1_all_quartiles_meta_AgeHG002.pdf"),
  plot = p,
  scale = 1,
  width = 10,
  height = 10,
)
#scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1),
#    labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) 
```

```{r}
# histogram of fraction of elements active at each age

frac.active <- rm %>%
  filter(!(abs(perc_sub - median(perc_sub)) > 2*sd(perc_sub))) %>%
  group_by(perc_sub) %>%
  summarise(frac = sum(status == "transcribed")/n())

hist <- ggplot(frac.active, aes(x=perc_sub, y=frac))+geom_histogram(stat="identity", position = "dodge")+coord_flip()+xlim(0,40)
```

```{r}


flankn <- 1000

n_windows=100
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
 # filter(quartile != 0) %>%
  group_by(cut,perc_sub) %>%
  mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency)) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  group_by(perc_sub) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 4, NA))


plot <- ggplot(chm13.ovl.labs,aes(x=min,y=perc_sub,fill=med_smooth))+ geom_tile()+scale_fill_gradientn(colours = rev(rainbow(5)))+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) +geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+theme(legend.position = "left", legend.direction="vertical")+ylim(0,40)

plot_grid( plot,hist,ncol=2, align = "v")

ggsave(
  paste0(figs, "/evol_meth/TE/","L1_methAge_heatmapSmoothHG002.pdf"),
  plot = last_plot(),
  scale = 1,
  width = 15,
  height = 10,
)
```


```{r}

flankn <- 1000

bodylen <- 300/1000
l1_regions <- rm %>%
  mutate(start = start - flankn, end = end + flankn) %>%
  GRanges()

ovl <- findOverlaps(hg002_gc, l1_regions)
genes.ovl <- as.data.frame(rm)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl <- as.data.frame(hg002_gc[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  dplyr::mutate(dist = ifelse(comp == "+",start - gene_start, gene_end - start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    )

n_windows=1050
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
 # filter(quartile != 0) %>%
  group_by(cut) %>%
  mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency)) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
 # group_by(quartile) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 30, NA))


# 4 is lowest divergence (youngest) 1 is highest divergence (oldest)
p <- ggplot(chm13.ovl.labs,aes( x = min, y = med_smooth), alpha=.5)+
#  geom_smooth(method = "loess", span = 0.12,se = F)+
#  geom_smooth(se=T)+lims( y = c(0,1)) +
 # geom_ribbon(aes(ymin=bot, ymax=top), alpha=0.2, size=0)+
  geom_line(aes(y=med_smooth), alpha=.5, size=1) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb")))+ylim(.15,.25)

#+
#geom_text(data = chm13.ovl %>% 
#              ungroup() %>%
#              select(ID,rep_type) %>%
#              distinct() %>%
#              group_by(rep_type) %>%
#              summarise(n_genes=n()) %>%
#              distinct() %>%
#              mutate(height=case_when(rep_type == "AluY" ~ 0, 
#                          rep_type == "AluJb" ~ .01, 
#                          TRUE ~ .02)),
#            aes(x=c(2),
#                y=.2+height,
#                label=paste("# repeats =",comma(n_genes))
#                ), hjust=1)+facet_wrap(~rep_type)
p

ggsave(
  paste0(figs, "/evol_meth/TE/","L1_meta_age_HG002GC_noquartile.pdf"),
  plot = p,
  scale = 1,
  width = 10,
  height = 5,
)
#scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1),
#    labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) 
```

```{r}


flankn <- 1000

n_windows=100
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
 # mutate(reg = ifelse(Name %in% CT.genes$Name, "CT", "non-CT")) %>%
 # filter(quartile != 0) %>%
  group_by(cut,perc_sub) %>%
  mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency)) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  group_by(perc_sub) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 10, NA))


plot <- ggplot(chm13.ovl.labs,aes(x=min,y=perc_sub,fill=med_smooth))+ geom_tile()+scale_fill_gradientn(colours = rev(rainbow(5)))+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb"))) +geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+theme(legend.position = "left", legend.direction="vertical")+ylim(0,40)

plot_grid( plot,hist,ncol=2, align = "v")

ggsave(
  paste0(figs, "/evol_meth/TE/","L1_methAge_heatmapSmoothHG002GC.pdf"),
  plot = last_plot(),
  scale = 1,
  width = 15,
  height = 10,
)
```
