---
title: "Untitled"
author: "G$"
date: "1/27/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("~/projects/chm13_meth/scripts/v1.0_final_assembly/utils/ilee_plot_utils.R")
library("ggsci")
source("~/projects/chm13_meth/scripts/v1.0_final_assembly/utils/methylation_R_utils.R")
library(tidyverse)
library(GenomicRanges)
library(zoo)
library(rtracklayer)

figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"

```


```{r cars}
exp <- read_tsv(paste0(dat, "/chm13_final_beds/chm13.CATv1.gene.tsv")) %>%
  dplyr::rename("gene_id" = gene)
 
chm13_meth <- read_tsv(paste0(dat, "/methylation_calls/methylation_frequency_50kb_split.tsv")) %>%
  GRanges()

reps <- read_tsv(paste0(dat, "/annotations/chm13.draft_v1.0.fasta_rm.bed"), col_names = c("chr", "start", "end", "rep", "len", "strand", "rep_type", "rep_fam")) %>%
  GRanges()

list=c("chr13", "chr14", "chr15", "chr21", "chr22")

cen <- read_tsv(paste0(dat, "/chm13_final_beds/cenRegions.bed"), col_names = c("chr", "start", "end","name"))%>%
  GRanges()

gffRangedData<-import.gff(paste0(dat, "/annotations/CHM13.gff3.gz"))
cat.gtf <-as(gffRangedData, "GRanges")

select=c("gene", "exon", "TSS")
genes <- as.data.frame(cat.gtf[(elementMetadata(cat.gtf)[, "type"] == "gene")])
feat <- as.data.frame(cat.gtf[(elementMetadata(cat.gtf)[, "type"] %in% select)])
ggplot( data=merge(genes, exp, by = "gene_id"),aes( x=TPM))+geom_density()+scale_x_log10()
genes.exp <- merge(genes, exp, by = "gene_id") %>%
  mutate(exp=case_when(TPM == 0 ~ "unexpressed", 
                   TRUE ~ "expressed")) %>%
  group_by(exp) %>%
  mutate(quartile=ntile(TPM,3)) %>%
  ungroup()%>%
  mutate(quartile = ifelse(exp == "unexpressed", 0, quartile)) %>%
  GRanges()

```

```{r}


junc <- read_tsv(paste0(dat,"/annotations/t2t-chm13.v1.0.cenSat_regions.bed"), col_names = c("chr", "start", "end", "region")) %>%
  GRanges()

keepi <- findOverlaps(genes.exp,junc)
freq.matched <- genes.exp[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
  mcols(freq.matched),
  mcols(junc[subjectHits(keepi)]))

junc.genes <- as.data.frame(freq.matched)%>%
  mutate(gene_start=start, gene_end=end) %>%
  mutate(direction = strand)

flankn <- 10000

bodylen <- 1
l1_regions <- junc.genes  %>%
  mutate(start = gene_start - flankn, end = gene_end + flankn) %>%
  GRanges()

ovl <- findOverlaps(chm13_meth, l1_regions)
genes.ovl <- junc.genes [subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = gene_end - gene_start) 

chm13.ovl <- as.data.frame(chm13_meth[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  dplyr::mutate(dist = ifelse(direction == "+",start...2 - gene_start, gene_end - start...2),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    ) %>%
  #mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  ungroup()
  


n_windows=1000
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
  group_by(cut,quartile) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency),quartile=quartile, region=region) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  distinct() %>%
  group_by(quartile) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 20, NA))


p <- ggplot(chm13.ovl.labs,aes( x = min, y = med_smooth, color=as.factor(quartile)))+
#  geom_smooth(method = "loess", span = 0.15,se = F,alpha=.1)+
#  geom_smooth(se=T)+lims( y = c(0,1)) +
#  geom_ribbon(aes(ymin=bot, ymax=top), alpha=0.2, size=0)+
  geom_line(aes(y=med_smooth), alpha=.5, size=1) +
  geom_point(aes(y=med_smooth), alpha=.05, size=1)+
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb")))+ylim(0,1)+theme(legend.position = "bottom")
p

# methylation sig:

```

```{r}


junc <- read_tsv(paste0(dat,"/annotations/t2t-chm13.v1.0.cenSat_regions.bed"), col_names = c("chr", "start", "end", "region")) %>%
  GRanges()

keepi <- findOverlaps(genes.exp,junc)
freq.matched <- genes.exp[-queryHits(keepi)]

#mcols(freq.matched) <- cbind.data.frame(
#  mcols(freq.matched),
#  mcols(junc[subjectHits(keepi)]))

junc.genes <- as.data.frame(freq.matched)%>%
  mutate(gene_start=start, gene_end=end) %>%
  mutate(direction = strand)

flankn <- 10000

bodylen <- 1
l1_regions <- junc.genes  %>%
  mutate(start = gene_start - flankn, end = gene_end + flankn) %>%
  GRanges()

ovl <- findOverlaps(chm13_meth, l1_regions)
genes.ovl <- junc.genes [subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = gene_end - gene_start) 

chm13.ovl <- as.data.frame(chm13_meth[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  dplyr::mutate(dist = ifelse(direction == "+",start...2 - gene_start, gene_end - start...2),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    ) %>%
  #mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  ungroup()
  


n_windows=1000
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
  group_by(cut,quartile) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency),quartile=quartile) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  distinct() %>%
  group_by(quartile) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 20, NA))


p <- ggplot(chm13.ovl.labs,aes( x = min, y = med_smooth, color=as.factor(quartile)))+
#  geom_smooth(method = "loess", span = 0.15,se = F,alpha=.1)+
#  geom_smooth(se=T)+lims( y = c(0,1)) +
#  geom_ribbon(aes(ymin=bot, ymax=top), alpha=0.2, size=0)+
  geom_line(aes(y=med_smooth), alpha=.5, size=1) +
  geom_point(aes(y=med_smooth), alpha=.05, size=1)+
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb")))+ylim(0,1)+theme(legend.position = "bottom")
p

# methylation sig:

```
```{r}

junc <- read_tsv(paste0(dat,"/annotations/t2t-chm13.v1.0.cenSat_regions.bed"), col_names = c("chr", "start", "end", "region")) %>%
  GRanges()

cgi <- read_tsv(paste0(dat, "/CGI/chm13.draft_v1.0_EMBOSS_CGI.gff"), col_names = c("chr", "prog", "feat", "start", "end")) %>%
  GRanges()

keepi <- findOverlaps(cgi,junc)
freq.matched <- cgi[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
  mcols(freq.matched),
  mcols(junc[subjectHits(keepi)]))

cgi.cen <- as.data.frame(freq.matched)
freq.matched <- cgi[-queryHits(keepi)]
cgi.nocen <- as.data.frame(freq.matched) %>%
  mutate(region="noncen")

cgi <- rbind(cgi.cen,cgi.nocen) %>%
  GRanges()

keepi <- findOverlaps(cgi,reps)
freq.matched <- cgi[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
  mcols(freq.matched),
  mcols(reps[subjectHits(keepi)]))

rep.cgi <- as.data.frame(freq.matched) %>%
  dplyr::select(c(seqnames, start, end, rep, rep_type,region)) %>%
  dplyr::rename("name"=rep) %>%
  dplyr::rename("element" =rep_type ) %>%
  distinct()

genes.gr <- feat %>%
  mutate(start=ifelse(type == "gene", start-2000,start)) %>%
  mutate(end=ifelse(type == "gene", start+2000,end)) %>%
  GRanges()
keepi <- findOverlaps(cgi,genes.gr)
freq.matched <- cgi[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
  mcols(freq.matched),
  mcols(genes.gr[subjectHits(keepi)]))

rep.genes <- as.data.frame(freq.matched) %>%
  dplyr::select(c(seqnames, start, end, source_gene_common_name,type,region)) %>%
  dplyr::rename("name"=source_gene_common_name) %>%
  dplyr::rename("element" = type) %>%
  distinct()

keepi <- findOverlaps(cgi,reps)
cgi.masked <- cgi[-queryHits(keepi)]
keepi <- findOverlaps(cgi.masked,genes.gr)
cgi.oulier <- cgi.masked[-queryHits(keepi)]

cgi.none <- as.data.frame(cgi.oulier) %>%
  dplyr::select(c(seqnames, start, end,region)) %>%
  mutate(name="none") %>%
  mutate(element = "none") %>%
  distinct()

cgi.all <- rbind(cgi.none,rep.genes,rep.cgi) %>%
  mutate(len=end-start)

```

```{r}
chr.stats <- as.data.frame(cgi) %>%
  group_by(seqnames, region) %>%
  summarize(total_cgi=sum(n())) 

stats <- cgi.all %>%
  mutate(reg=ifelse(region=="noncen", "noncen", "cen")) %>%
  group_by(seqnames, reg, element) %>%
  summarize(cgi=sum(n()))

ggplot(stats, aes(x=seqnames, y=cgi, fill=element))+geom_bar(stat="identity")+facet_wrap(~reg, scales = "free", ncol=1)

```

