---
title: "SD Methylation"
author: "Ariel"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  pdf_document: default
  html_document:
    df_print: paged
  html_notebook: default
mainfont: DejaVu Sans
titlefont: DejaVu Sans
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE,  message=FALSE)
library(knitr)
library(tidyverse)
library(wesanderson)
library(BSgenome)
library(png)
library(cowplot)
options(scipen=999)
library(zoo)
library(ggridges)
library(GenomicRanges)
library(RColorBrewer)
library(BSgenome.t2t.v1.0.release)
options(knitr.duplicate.label = 'allow')
source("/home/isac/Code/ilee/plot/ilee_plot_utils.R")
library("ggsci")
library(gridExtra) 
library(pheatmap)
source("/home/isac/Code/nanopore-methylation-utilities/methylation_R_utils.R")
```

```{r loadData, include=F}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/sd/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"

flank <- read_tsv(paste0(dat, "/sd/blocks.with.flanks.nowindows.bed"), col_names = c("seqnames", "start", "end")) %>%
  mutate(ID = row_number())
  
chm13_meth <- read_tsv(paste0(dat, "/methylation_calls/methylation_frequency_50kb_split.tsv"))
chm13_CpG <- readRDS(paste0(dat, "/reference/chm13_sliding_200_CpG.rds"))
```


```{r functions, include=F}

ovls<-function(gr1, gr2){
  keepi <- findOverlaps(gr1, gr2)
freq.matched <- gr1[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
    mcols(freq.matched),
    mcols(gr2[subjectHits(keepi)]))

df <- as.data.frame(freq.matched)
return(df)
}

```

# Methylation within and 50kb flanking SDs 
Criteria for SDs included in this analysis:
- SDs > 50kb in length 
- Do not have another SD within 50kb of flanking regions 
- Comparing methylation in normalized SD length and +/-50kb of the flanking regions for genes 
- For heatmaps each row is an individual SD 
- Aggregated methylation plot is the smoothed average of all SDs
```{r SD_all, include=F}

# SDs all 

flankn <- 20e3
bodylen <- 3
SD_regions <- flank %>%
  mutate(
    start = start - flankn,
    end = end + flankn) %>%
  GRanges()

# find overlaps between methylation calls and SDs
chm13_meth <- GRanges(chm13_meth)
ovl <- findOverlaps(chm13_meth, SD_regions)
genes.ovl <- as.data.frame(flank)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

# put all SDs on same X axis coordinates
chm13.ovl <- as.data.frame(chm13_meth[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  mutate(dist = start - gene_start) %>%
  mutate(dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,2)
    )


# write table for PerBase methylation in SDs
write.table(chm13.ovl, file = paste0(dat, "/sd/t2t_chm13v1.0_SD_clustered_methylationPerBase.bed"),quote = F, sep = "\t", row.names = F, col.names = T)

# make distance labels
chm13.ovl.labs <- chm13.ovl %>%
  group_by(dist) %>%
  summarise(med_meth = median(methylated_frequency)) %>%
  distinct()


# cluster and make distance matrix
clust_meth <- chm13.ovl %>% 
  select(c(ID,methylated_frequency,dist)) %>%
  group_by(ID, dist) %>%
  summarise(meth = mean(methylated_frequency)) %>%
  spread(dist,meth) %>%
  column_to_rownames("ID") %>%
  as.matrix() 

# annotation of flank vs SD for plot
annot <- chm13.ovl %>% 
  select(c(ID,methylated_frequency,dist)) %>%
  group_by(ID, dist) %>%
  summarise(meth = mean(methylated_frequency))%>%
  mutate(reg = case_when( dist < 0 ~ "flank", 
                    dist > bodylen ~ "flank", 
                    TRUE ~ "SD")) %>%
  ungroup() %>%
  select(c(dist, reg)) %>%
  distinct() %>%
  column_to_rownames("dist")


```

```{r, echo=F, fig.height=5, fig.width=8}

# make heatmap and clustered bed
library(viridis)

res <- pheatmap(clust_meth, cluster_cols = F, annotation_col = annot, show_rownames = F, show_colnames = F, main = "SD Methylation")
res

SDmeth.clust <- as.data.frame(clust_meth) %>%
  mutate(avgmeth = rowMeans(., na.rm = T)) %>%
  mutate(cluster_num = cutree(res$tree_row, k = 2)) %>%
  rownames_to_column("ID") %>%
  select(c(ID, cluster_num, avgmeth)) %>%
  mutate(clust_meth = case_when(cluster_num == 1 ~ "unmeth", 
                                cluster_num == 2 ~ "meth"))
  
final_clust <- merge(flank,SDmeth.clust, by = "ID") %>%
  select(-c(ID)) %>%
  rename(seqnames = "chr")

write.table(final_clust, file = paste0(dat, "/sd/t2t_chm13v1.0_SD_clustered_methylation.bed"),quote = F, sep = "\t", row.names = F, col.names = T)

ggsave(
  paste0(figs, "/","SD_methylation_heatmap.pdf"),
  plot = res,
  scale = 1,
  width = 10,
  height = 10,
)

```

