---
title: "Composite Element Methylation TELOclusters"
author: "Ariel"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  pdf_document: default
  html_document:
    df_print: paged
  html_notebook: default
mainfont: DejaVu Sans
titlefont: DejaVu Sans
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE,  message=FALSE)
source("~/T2T-Epigenetics/utils/ilee_plot_utils.R")
source("~/T2T-Epigenetics/utils/methylation_R_utils.R")
library(tidyverse)
library(cowplot)

figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"
```

```{r loadData, include=F}
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"
chm13_meth <- read_tsv(paste0(dat, "/methylation_calls/methylation_frequency_50kb_split.tsv")) %>%
  GRanges()

regs <- read_tsv(paste0(dat, "/TE/TeloComposites/TELO-composite_RM-polish_APR0521_SHoyt.txt"), col_names = T) %>%
  mutate(len=end-start) %>%
  filter(len>0) %>%
  filter(len<3000) 

```

```{r 5kb, echo=F}


flankn <- 20000
bodylen <- 1/5

#flip = c("chr5", "chr9", "chr3", "chr16")
l1_regions <- regs %>%
  mutate(start = start - flankn, end = end + flankn) %>%
  GRanges()

ovl <- findOverlaps(chm13_meth, l1_regions)
genes.ovl <- as.data.frame(regs)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl <- as.data.frame(chm13_meth[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  mutate(dist = start - gene_start) %>%
  mutate(dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,2)
    )


#chm13.stat <- chm13.ovl %>%
#  group_by(dist, subTELO) %>%
#  summarise(meth=mean(methylated_frequency))

n_windows=500
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
  group_by(cut,subTELO) %>%
 # mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency)) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  group_by(subTELO) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 10, NA))

p <- ggplot(chm13.ovl.labs,aes( x = min, y = med_smooth, color=subTELO, fill=subTELO), alpha=.5)+
#  geom_smooth(method = "loess", span = 0.12,se = F)+
#  geom_smooth(se=T)+lims( y = c(0,1)) +
  geom_ribbon(aes(ymin=bot, ymax=top), alpha=0.2, size=0)+
  geom_line(aes(y=med_smooth), alpha=.5, size=1) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb")))+
geom_text(data = chm13.ovl %>% 
              group_by(subTELO) %>%
              select(subTELO,gene_start) %>%
              distinct() %>%
              summarise(n_genes=n()) %>%
              distinct() %>%
              mutate(height=case_when(subTELO == "subTELO" ~ 0, 
                          subTELO == "no" ~ .1, 
                          TRUE ~ .02)),
            aes(x=c(2),
                y=.5+height,
                label=paste("n =",comma(n_genes))
                ), hjust=1)

p


ggsave(
  paste0(figs, "/evol_meth/TE/","CHM13_subTELO_metaplot.pdf"),
  plot = p,
  scale = 1,
  width = 12,
  height = 5,
)
```












