---
title: "201008_TE_methylation"
author: "G$"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  pdf_document: default
  html_notebook: default
mainfont: DejaVu Sans
titlefont: DejaVu Sans
---



```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(wesanderson)
library(ggExtra)
library(biovizBase)
library(Repitools)
library(Biostrings)
library(ggplot2)
library(BSgenome)
library(png)
library(cowplot)
options(scipen=999)
library(zoo)
library(ggridges)
library(GenomicRanges)
library(RColorBrewer)
library(BSgenome.t2t.v1.1)
options(knitr.duplicate.label = 'allow')
source("/home/isac/Code/ilee/plot/ilee_plot_utils.R")
library("ggsci")
library(gridExtra) 
source("/home/isac/Code/nanopore-methylation-utilities/methylation_R_utils.R")
```

```{r loadData, include=F}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/final_assembly/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/final_assembly"

#gffRangedData<-import.gff(paste0(dat, "/liftoff/hg38_genes_t2t_liftoff.gtf")) 
chm13 <- readRDS(paste0(dat, "/ref/chm13_200bpCpG_cov_200_CpG_num.rds"))

reps <- read_tsv(paste0(dat, "/annotations/t2t-chm13.20200727_repeatmasker.out.bed") ,col_names= F) %>%
  dplyr::select(X1, X2, X3, X4, X7, X8) %>%
  dplyr::rename("chr" = X1, "start" = X2, "end" = X3, "rep_type" = X7, "sub_type" = X8, "ev_name" = X4) %>%
  mutate(len = end - start) %>%
  mutate(ID = row_number())
SD <- read_tsv(paste0(dat, "/annotations/t2t-chm13.20200727.SDs.bed")) %>%
  dplyr::rename("chr" = 1, "start" = 2, "end" = 3) %>%
  mutate(start = start +1)

cens <-  read_tsv(paste0(dat, "/annotations/t2t-chm13.20200727.cenRegions.bed"), col_names = F)  %>%
  dplyr::rename("chr" = 1, "start" = 2, "end" = 3) %>%
  mutate(start = start +1)
censat = read_tsv(paste0(dat, "/annotations/t2t-chm13.20200727.cenSat_annotation.bed"), col_names = F) %>%
  mutate(name = ifelse(grepl("HSat1", X4), "HSAT1", X4)) %>%
  mutate(name = ifelse(grepl("HSat2", X4), "HSAT2", name)) %>%
  mutate(name = ifelse(grepl("HSat3", X4), "HSAT3", name)) %>%
  mutate(name = ifelse(grepl("HSat4", X4), "HSAT4", name)) %>%
  mutate(name = ifelse(grepl("HSat5", X4), "HSAT5", name)) %>%
  mutate(name = ifelse(grepl("HSat6", X4), "HSAT6", name)) %>%
  mutate(name = ifelse(grepl("ct", X4), "CT", name)) %>%
  mutate(name = ifelse(grepl("bsat", X4), "BSAT", name)) %>%
  mutate(name = ifelse(grepl("hor", X4), "HOR", name)) %>%
  mutate(name = ifelse(grepl("mon", X4), "MON", name)) %>%
  mutate(name = ifelse(grepl("Alu", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("SATR", X4), "SATR", name)) %>%
  mutate(name = ifelse(grepl("ACRO", X4), "ACRO", name)) %>%
  mutate(name = ifelse(grepl("GSATII", X4), "GSATII", name)) %>%
  mutate(name = ifelse(grepl("TAR", X4), "TAR", name)) %>%
  mutate(name = ifelse(grepl("TE", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("MER", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("MST", X4), "MST", name)) %>%
  mutate(name = ifelse(grepl("CER", X4), "CER", name)) %>%
  mutate(name = ifelse(grepl("L1", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("SST", X4), "SST", name)) %>%
  mutate(name = ifelse(grepl("LSAU", X4), "LSAU", name)) %>%
  mutate(name = ifelse(grepl("GSAT", X4), "GSAT", name)) %>%
  mutate(name = ifelse(grepl("MSAT", X4), "MSAT", name)) %>%
  mutate(name = ifelse(grepl("novel", X4), "novel", name)) %>%
  mutate(name = ifelse(grepl("HERV", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("LT", X4), "TE", name)) %>%
  dplyr::select(c(X1, X2, X3, name)) %>%
  dplyr::rename("chrom" =1, "start" = 2 ,"end" =3) %>%
  ungroup()
table(censat$name)

age <-  read_tsv(paste0(dat, "/annotations/RM_L1_evage.bed"), col_names = c("ev_name", "age"))
```

```{r functions, include=F}

CalculateCpG <- function (obj, ..., step, width, as.prob, as.array, fast.moving.side,with.labels) 
{
  require(BSgenome)
  seqs <- getSeq(obj, ..., as.character = FALSE)
  if (missing(step)) {
    res <- oligonucleotideFrequency(seqs, 2, as.prob = as.prob)[,7]
  }
  else {
    res <- lapply(seqs, function(x) oligonucleotideFrequency(seqs, width = 2, step = step, as.prob = as.prob, as.array = as.array,fast.moving.side=fast.moving.side,with.labels=with.labels)[,7])
    if (length(res) == 1) 
      res <- unlist(res)
  }
  res
}


ovls<-function(gr1, gr2){
  keepi <- findOverlaps(gr1, gr2)
freq.matched <- gr1[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
    mcols(freq.matched),
    mcols(gr2[subjectHits(keepi)]))

df <- as.data.frame(freq.matched)
return(df)
}
```



# LINE elements
```{r}

LINE <- reps %>%
  filter(sub_type == "L1")

LINE <- merge(LINE, age, by = "ev_name")

LINE <- GRanges(LINE)
flankn <- 1e3
bodylen <- 2
regions <- as.data.frame(LINE) %>%
  mutate( 
    start = start - flankn,
    end = end + flankn) %>%
  GRanges()

ovl <- findOverlaps(chm13, regions)
genes.ovl <- as.data.frame(LINE)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end, gene_strand = strand)

chm13.ovl <- as.data.frame(chm13[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  mutate(dist = ifelse(gene_strand == "-",gene_end - start, start - gene_start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    )

#chm13.ovl.labs <- chm13.ovl %>% 
#  group_by(age) %>% 
#  dplyr::arrange(dist) %>% 
#  dplyr::mutate(Loess = predict(loess(meth ~ dist, span = .5, data=.)))
#

chm13.ovl.labs <- chm13.ovl %>%
  group_by(age, dist) %>%
  summarise(med_meth = median(meth)) %>%
  ungroup() %>%
  group_by(age) %>% 
  arrange(dist)
  
 # mutate(Loess = predict(loess(med_meth ~ dist, span = .015)))
library(viridis)

p <- ggplot(chm13.ovl.labs,aes( x = dist, y = as.factor(age))) +
  geom_tile(aes(fill = med_meth)) + 
 scale_fill_distiller(palette = "YlGnBu", direction = 1,limits=c(0,1))+geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen) +
  scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1),
    labels = c(paste0("-",flankn/1e3,"kb"),"LINE start","LINE end",paste0("+",flankn/1e3,"kb"))) +
  labs( x = "Genomic Position", y = "Evolutionary Age (Myr") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()

p
```


```{r functions, include=F}
CalculateCpG <- function (obj, ..., step, as.prob = TRUE) 
{
    require(BSgenome)
    seqs <- getSeq(obj, ..., as.character = FALSE)
    if (missing(step)) {
        res <- oligonucleotideFrequency(seqs, 2, as.prob = as.prob)[,7]
    }
    else {
        res <- lapply(seqs, function(x) oligonucleotideFrequency(seqs, 2, step = step, as.prob = as.prob)[,7])
        if (length(res) == 1) 
            res <- unlist(res)
    }
    res
}


g_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend) }

smoothplot<-function(cen){
pal <- wes_palette("Zissou1", 200, type = "continuous")

cen <- cen %>%
  mutate(cpgsmooth = rollmean(CpG, 200, fill = NA)) %>%
  mutate(methsmooth = rollmean(meth, 200, fill = NA))

pal <- wes_palette("Zissou1", 200, type = "continuous")
p <- ggplot(cen, aes(x=cpgsmooth, y = methsmooth))+
  stat_density_2d(aes(fill = ..count..), geom = "raster", contour = F)+scale_fill_gradientn(colours = pal)+geom_density_2d(color = "black", linetype = "dotted")+xlim(0,20)+theme(legend.position = c(0.8, 0.2))+theme_classic()+ylim(0,1)

thm = list(theme_void(),
           guides(fill=FALSE),
           theme(plot.margin=unit(rep(2.5,4), "lines")))
leg = g_legend(p)
p1 = p + thm[-1]
p2 = ggplot(cen, aes(meth)) +
  geom_histogram() +
  thm +
  coord_cartesian(0,1)+
  coord_flip()
p3 = ggplot(cen, aes(CpG)) +
  geom_histogram(bins=100) + 
  thm +xlim(0,20)
plot_final <- plot_grid(
  plot_grid(plotlist=list(p3, ggplot(), p1, p2), ncol=2, 
            rel_widths=c(4,1), rel_heights=c(1,4), align="hv", scale=1.1),leg, rel_widths=c(5,1))

return(plot_final)
}


prettyplot<-function(cen){
pal <- wes_palette("Zissou1", 200, type = "continuous")

cen <- cen %>%
  mutate(smooth = rollmean(CpG, 200, fill = NA))

p <- ggplot(cen, aes(x=smooth, y = meth))+
  geom_bin2d(aes(fill=..count..)) +
  scale_fill_gradientn(colours = pal)+
  theme_classic()+xlim(0,20)+theme(legend.position = c(0.8, 0.2))
thm = list(theme_void(),
           guides(fill=FALSE),
           theme(plot.margin=unit(rep(2.5,4), "lines")))
leg = g_legend(p)
p1 = p + thm[-1]
p2 = ggplot(cen, aes(meth)) +
  geom_histogram() +
  thm +
  coord_cartesian(0,1)+
  coord_flip()
p3 = ggplot(cen, aes(smooth)) +
  geom_histogram(bins=100) + 
  thm +xlim(0,20)
plot_final <- plot_grid(
  plot_grid(plotlist=list(p3, ggplot(), p1, p2), ncol=2, 
            rel_widths=c(4,1), rel_heights=c(1,4), align="hv", scale=1.1),leg, rel_widths=c(5,1))

return(plot_final)
}


ovls<-function(gr1, gr2){
  keepi <- findOverlaps(gr1, gr2)
freq.matched <- gr1[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
    mcols(freq.matched),
    mcols(gr2[subjectHits(keepi)]))

df <- as.data.frame(freq.matched)
return(df)
}

```

```{r metaplots,include=F}

#genes <- gffRangedData[gffRangedData$type == "gene" ]

cen.gr <- GRanges(cens)
type = c("LINE", "SINE", "LTR")
SINE <- reps %>%
  filter(rep_type %in% type) 


SINE_CT <- ovls(cen.gr, GRanges(SINE)) %>%
  select(ID) %>%
  distinct()

SINE <- SINE %>%
  mutate(reg = ifelse(ID %in% SINE_CT$ID, "CT", "Non-CT")) %>%
  mutate(strand = "*") 
                                                                                                    
flankn <- 1e3
bodylen <- 2
regions <- as.data.frame(SINE) %>%
  mutate( 
    start = start - flankn,
    end = end + flankn) %>%
  GRanges()

ovl <- findOverlaps(chm13, regions)
genes.ovl <- as.data.frame(SINE)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end, gene_strand = strand) 

chm13.ovl <- as.data.frame(chm13[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  mutate(dist = ifelse(gene_strand == "*",gene_end - start, start - gene_start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    )

#CT.gr <- GRanges(censat %>%
#  filter(name == "CT"))


#CT.genes <- as.data.frame(ovls(regions,CT.gr)) %>%
#  select(Name) %>%
#  distinct()

chm13.ovl.labs <- chm13.ovl %>%
  group_by(dist, reg, rep_type) %>%
  summarise(med_meth = median(meth)) %>%
  distinct()

# methylation in CT genes vs all other genes 
p <- ggplot(chm13.ovl.labs,aes( x = dist, y = med_meth, color = reg)) +
  geom_smooth(method = "loess", span = 0.15,se = F) + lims( y = c(0,1)) +
#  geom_smooth(se=F)+lims( y = c(0,1)) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen) +
  scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1),
    labels = c(paste0("-",flankn/1e3,"kb"),"TSS","TTS",paste0("+",flankn/1e3,"kb"))) +
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+facet_wrap(~rep_type, scales = "free")
p

#ggsave(
#  paste0(figs, "/","gene_methylation_metaplot.pdf"),
#  plot = p,
#  scale = 1,
#  width = 10,
#  height = 10,
#)
#

```

# Alu elements
```{r alu, echo=F}

# SDs all 

cen.gr <- GRanges(cens)

SINE <- reps %>%
  filter(rep_type == "LINE") 


SINE_CT <- ovls(cen.gr, GRanges(SINE))
  
row_annot <- SINE_CT %>%
    select(ID, sub_type) %>%
    distinct() %>%
    column_to_rownames("ID")
  
#body_CG <- CalculateCpG(BSgenome.t2t.v1.1, GRanges(SINE_CT), as.prob = F)
#alus <- as.data.frame(GRanges(GRanges(SINE_CT), CpG = body_CG, ID = SINE_CT$ID, len = SINE_CT$len)) %>%
#  mutate(CpG_freq = CpG/len) %>%
#  arrange(CpG_freq) %>%
#  mutate(order = row_number()) %>%
#  select(c(ID, order))
#
meth_order <- ovls(chm13, GRanges(SINE_CT)) %>%
  group_by(ID) %>%
  summarise(meth = mean(meth)) %>%
  arrange(meth) %>%
  mutate(order = row_number()) %>%
  select(c(order,ID))


flankn <- 1e3
bodylen <- 1
alu_regions <- SINE_CT %>%
  mutate(
    start = start - flankn,
    end = end + flankn) %>%
  GRanges()

ovl <- findOverlaps(chm13, alu_regions)
genes.ovl <- as.data.frame(SINE_CT)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl <- as.data.frame(chm13[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  mutate(dist = start - gene_start) %>%
  mutate(dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,2)
    )


chm13.ovl.labs <- chm13.ovl %>%
  group_by(dist) %>%
  summarise(med_meth = median(meth)) %>%
  distinct()


# cluster and make distance matrix
#alu.ordered <- merge(alus, chm13.ovl, by = "ID")

meth <- chm13.ovl %>% 
  select(c(ID, meth,dist)) %>%
  group_by(ID, dist) %>%
  mutate(meth = mean(meth)) %>%
  distinct() %>%
  spread(ID, meth) %>%
  column_to_rownames("dist")

smooth <- meth %>%
  na.approx(rule = 2) %>%
  as.data.frame() %>%
  dplyr::mutate(row = rownames(meth)) %>%
  column_to_rownames("row") %>%
  t() %>%
  na.omit() %>%
  as.matrix()

#na.gam.replace(as.data.frame(clust_meth))

CG <- chm13.ovl %>% 
  select(c(ID,CpG,dist)) %>%
  group_by(ID, dist) %>%
  mutate(CpG = mean(CpG)) %>%
  distinct() %>%
  spread(ID,CpG) %>%
  column_to_rownames("dist")

smooth_CG <- CG %>%
  na.approx(rule = 2) %>%
  as.data.frame() %>%
  dplyr::mutate(row = rownames(CG)) %>%
  column_to_rownames("row") %>%
  t() %>%
  as.matrix()

write.table(smooth, file = paste0(dat,"/LINE_meth_smooth.tsv"), quote = F, sep = "\t", row.names = F,
            col.names = F, qmethod = c("escape", "double"),
            fileEncoding = "")
#  clust_CpG <- alu.ordered %>% 
#  select(c(order,CpG,dist)) %>%
#  group_by(order, dist) %>%
#  summarise(CG = mean(CpG)) %>%
#  spread(dist,CG) %>%
#  column_to_rownames("order") %>%
#  as.matrix() 

library(gplots)
library(pheatmap)
coul <- colorRampPalette(brewer.pal(8, "PiYG"))(25)

quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(xs, probs = seq(0, 1, length.out = n), na.rm = T)
  breaks[!duplicated(breaks)]
}


annot <- chm13.ovl %>% 
  select(c(ID,meth,dist)) %>%
  group_by(ID, dist) %>%
  summarise(meth = mean(meth))%>%
  mutate(reg = case_when( dist < 0 ~ "flank", 
                    dist > bodylen ~ "flank", 
                    TRUE ~ "SD")) %>%
  ungroup() %>%
  select(c(dist, reg)) %>%
  distinct() %>%
  column_to_rownames("dist")
library(gam)



mat_breaksCG <- quantile_breaks(smooth_CG, n = 11)
#mat_breaksmeth <- quantile_breaks(clust_meth, n = 11)
library(viridis)
library(pheatmap)
p1 <- pheatmap(smooth_CG, cluster_cols = F,cluster_rows = T, annotation_col = annot, show_rownames = F, show_colnames = F, breaks= mat_breaksCG, color = inferno(length(mat_breaksCG) - 1), main = "LINE CpG Density in 200bp window", annotation_row = row_annot)

p2 <- pheatmap(smooth, cluster_cols = F,  cluster_rows = T,annotation_row = row_annot,show_rownames = F, annotation_col = annot, show_colnames = F, main = "LINE Methylation")

print(p1)
print(p2)

ggsave(
  paste0(figs, "/","LINE_CG_heatmap.pdf"),
  plot = p1,
  scale = 1,
  width = 10,
  height = 20,
)

ggsave(
  paste0(figs, "/","LINE_meth_heatmap.pdf"),
  plot = p2,
  scale = 1,
  width = 10,
  height = 20,
)
```


# Alu elements
```{r}

flankn <- 10e3
bodylen <- .5

LINE <- reps %>%
  filter(sub_type == "Alu") %>%
  slice_head(n = 1000) 

regions <- as.data.frame(LINE) %>%
  mutate( 
    alu_start = start, 
    alu_end = end, 
    start = alu_start - flankn,
    end = alu_end + flankn) %>%
  GRanges()


body_CG <- CalculateCpG(BSgenome.t2t.v1.1, GRanges(LINE), as.prob = F)
alus <- as.data.frame(GRanges(GRanges(LINE), CpG = body_CG, ID = LINE$ID, len = LINE$len)) %>%
  mutate(CpG_freq = CpG/len) %>%
  arrange(CpG_freq) %>%
  mutate(order = row_number()) %>%
  select(c(ID, order))


windows <- slidingWindows(regions,width=100L, step=1L)
windows <- unlist(windows)
seqs <- getSeq(BSgenome.t2t.v1.1, windows, as.character = FALSE)
res <- oligonucleotideFrequency(seqs, 2, as.prob =F)[,7]
alu_windows <- GRanges(windows, CpG = res)

alu_CpG <- ovls(alu_windows, GRanges(regions)) %>%
  mutate(dist = ifelse(strand == "*",alu_end - start, start - alu_start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < len,
        bodylen * dist / len,
        bodylen + (dist - len)/flankn)), 
    dist = round(dist,3)) %>%
      mutate(CpG_freq = CpG/len) %>%
  mutate(smooth = rollmean(CpG_freq, 20, fill = NA))
    
alu_CpG <- merge(alu_CpG, alus, by = "ID")

p <- ggplot(alu_CpG,aes( x = dist, y = order)) +
  geom_tile(aes(fill = CpG_freq)) + 
 scale_fill_distiller(palette = "YlGnBu", direction = 1,limits=c(0,.1))+geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen) +
  scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1),
    labels = c(paste0("-",flankn/1e3,"kb"),"LINE start","LINE end",paste0("+",flankn/1e3,"kb"))) +
  labs( x = "Genomic Position", y = "Evolutionary Age (Myr") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()

p


```