---
title: "0727_SD_meth"
author: "G$"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  pdf_document: default
  html_notebook: default
mainfont: DejaVu Sans
titlefont: DejaVu Sans
---
```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(Biostrings)
library(ggplot2)
library(png)
library(cowplot)
options(scipen=999)
library(zoo)
options(knitr.duplicate.label = 'allow')
source("/home/isac/Code/ilee/plot/ilee_plot_utils.R")
library("ggsci")
source("/home/isac/Code/nanopore-methylation-utilities/methylation_R_utils.R")
library(mclust)
library(ggridges)
```

```{r functions, include=F}
CalculateCpG <- function (obj, ..., step, as.prob = TRUE) 
{
    require(BSgenome)
    seqs <- getSeq(obj, ..., as.character = FALSE)
    if (missing(step)) {
        res <- oligonucleotideFrequency(seqs, 2, as.prob = as.prob)[,7]
    }
    else {
        res <- lapply(seqs, function(x) oligonucleotideFrequency(seqs, 2, step = step, as.prob = as.prob)[,7])
        if (length(res) == 1) 
            res <- unlist(res)
    }
    res
}


g_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend) }

smoothplot<-function(cen){
pal <- wes_palette("Zissou1", 200, type = "continuous")

cen <- cen %>%
  mutate(smooth = rollmean(CpG, 200, fill = NA)) %>%
  mutate(meth = rollmean(meth, 200, fill = NA))

pal <- wes_palette("Zissou1", 200, type = "continuous")
p <- ggplot(cen, aes(x=smooth, y = meth))+
  stat_density_2d(aes(fill = ..count..), geom = "raster", contour = F)+scale_fill_gradientn(colours = pal)+geom_density_2d(color = "black", linetype = "dotted")+xlim(0,15)+theme(legend.position = c(0.8, 0.2))+theme_classic()+ylim(0,1)

thm = list(theme_void(),
           guides(fill=FALSE),
           theme(plot.margin=unit(rep(2.5,4), "lines")))
leg = g_legend(p)
p1 = p + thm[-1]
p2 = ggplot(cen, aes(meth)) +
  geom_histogram() +
  thm +
  coord_cartesian(0,1)+
  coord_flip()
p3 = ggplot(cen, aes(smooth)) +
  geom_histogram(bins=100) + 
  thm +xlim(0,20)
plot_final <- plot_grid(
  plot_grid(plotlist=list(p3, ggplot(), p1, p2), ncol=2, 
            rel_widths=c(4,1), rel_heights=c(1,4), align="hv", scale=1.1),leg, rel_widths=c(5,1))

return(plot_final)
}


prettyplot<-function(cen){
pal <- wes_palette("Zissou1", 200, type = "continuous")

cen <- cen %>%
  mutate(smooth = rollmean(CpG, 200, fill = NA))

p <- ggplot(cen, aes(x=smooth, y = meth))+
  geom_bin2d(aes(fill=..count..)) +
  scale_fill_gradientn(colours = pal)+
  theme_classic()+xlim(0,20)+theme(legend.position = c(0.8, 0.2))
thm = list(theme_void(),
           guides(fill=FALSE),
           theme(plot.margin=unit(rep(2.5,4), "lines")))
leg = g_legend(p)
p1 = p + thm[-1]
p2 = ggplot(cen, aes(meth)) +
  geom_histogram() +
  thm +
  coord_cartesian(0,1)+
  coord_flip()
p3 = ggplot(cen, aes(smooth)) +
  geom_histogram(bins=100) + 
  thm +xlim(0,20)
plot_final <- plot_grid(
  plot_grid(plotlist=list(p3, ggplot(), p1, p2), ncol=2, 
            rel_widths=c(4,1), rel_heights=c(1,4), align="hv", scale=1.1),leg, rel_widths=c(5,1))

return(plot_final)
}


ovls<-function(gr1, gr2){
  keepi <- findOverlaps(gr1, gr2, type = "any")
freq.matched <- gr1[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
    mcols(freq.matched),
    mcols(gr2[subjectHits(keepi)]))

df <- as.data.frame(freq.matched)
return(df)
}



ovls_chunked <-function(gr1, gr2){
  keepi <- findOverlaps(gr1, gr2)
freq.matched <- gr1[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
    mcols(freq.matched),
    mcols(gr2[subjectHits(keepi)]))

df <- as.data.frame(freq.matched) %>%
  group_by(seqnames, start, end, width) %>%
  summarise(meth = mean(meth), 
            cov = mean(as.numeric(cov)), 
            CpG = mean(CpG))

return(df)
}

chunkplot <-function(cen){
pal <- wes_palette("Zissou1", 200, type = "continuous")

pal <- wes_palette("Zissou1", 200, type = "continuous")
p <- ggplot(cen, aes(x=CpG, y = meth))+
  stat_density_2d(aes(fill = ..count..), geom = "raster", contour = F)+scale_fill_gradientn(colours = pal)+geom_density_2d(color = "black", linetype = "dotted")+xlim(0,15)+theme(legend.position = "bottom")+theme_classic()+ylim(0,1)+theme(text = element_text(size = 30))+labs(x = "CpG density", y = "Methylation frequency")

thm = list(theme_void(),
           guides(fill=FALSE),
           theme(plot.margin=unit(rep(2.5,4), "lines")))
leg = g_legend(p)
p1 = p + thm[-1]
p2 = ggplot(cen, aes(meth)) +
  geom_histogram() +
  thm +
  coord_cartesian(0,1)+
  coord_flip()
p3 = ggplot(cen, aes(CpG)) +
  geom_histogram(bins=100) + 
  thm +xlim(0,20)
plot_final <- plot_grid(
  plot_grid(plotlist=list(p3, ggplot(), p1, p2), ncol=2, 
            rel_widths=c(4,1), rel_heights=c(1,4), align="hv", scale=1.1),leg, rel_widths=c(5,1))

return(plot_final)
}


order_reads_facet <- function(x,offset = 0, bounds=NULL, qorder = NULL){
  # get boundaries of reads if not provided
  if (is.null(bounds)){
    bounds <- x %>% group_by(qname, clust) %>%
      summarize(start = min(start),
                end = max(end))
    # label y based on order of smallest start
    # if q order is given, use this to order
    if (!is.null(qorder)) {
      bounds <- bounds %>%
        mutate(qname = factor(qname,levels = qorder)) %>%
        arrange(qname) 
    } else {
      bounds<- bounds %>% 
        arrange(start, end) 
    }
    bounds <- bounds %>%
      ungroup() %>%
      group_by(clust) %>%
        mutate(
          readi = row_number(),
          ymin = -readi - 0.8 - offset, 
          ymax = ymin + 0.6)
  }
  
  x <- x %>%
    mutate(ymin = bounds$ymin[match(qname,bounds$qname)],
           ymax = bounds$ymax[match(qname,bounds$qname)])
  bounds <- bounds %>%
    mutate(ymin = bounds$ymin[match(qname,bounds$qname)],
           ymax = bounds$ymax[match(qname,bounds$qname)])
  return(list(x = x,bounds = bounds))
}
```

```{r data, include=F}

figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/final_assembly/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/final_assembly"


reps <- read_tsv(paste0(dat, "/annotations/t2t-chm13.20200727_repeatmasker.out.bed") ,col_names= F) %>%
  select(X1, X2, X3, X4, X7, X8) %>%
  dplyr::rename("chr" = X1, "start" = X2, "end" = X3, "name" = X4, "rep_type" = X7, "sub_type" =X8) %>%
  mutate(len = end - start) %>%
  filter(sub_type == "L1") %>%
  mutate(ID = row_number())

chm13 <- readRDS(paste0(dat, "/ref/chm13_200bpCpG_cov_200_CpG_num.rds"))

sizes <- read_tsv(paste0(dat, "/ref/t2t-chm13.20200727.fasta.fai"), col_names = c("chr", "size"))
```

# Distribution of L1 elements across chromosomes 

```{r distro, echo=F}

rep.tot <- reps %>%
  group_by(chr) %>%
  summarize(total_bases = sum(len))

p2 <- merge(rep.tot, sizes, by = "chr") %>%
  group_by(chr) %>%
  summarize(total_L1 = total_bases/size) %>%
  ggplot(aes(chr, total_L1))+geom_bar(stat = "identity", position = "dodge")+labs(x = "Chromosome", y = "Fraction of total bases")
    
p2

```

```{r meth_distro, echo = F}

reps.gr <- GRanges(reps)

reps.meth <- ovls(chm13, reps.gr) %>%
  group_by(ID, seqnames, name) %>%
  summarise(meth = mean(meth))

ggplot(reps.meth, aes(seqnames,meth, fill = seqnames))+geom_violin()+facet_wrap(~name)


l1x <- reps.meth %>%
 # mutate(name = ifelse(grepl("HAL", name), "HAL", name)) %>%
 # mutate(name = ifelse(grepl("L1H", name), "L1H", name)) %>%
  mutate(name = ifelse(grepl("L1P", name), "L1P", name)) %>%
# mutate(name = ifelse(grepl("L1M4", name), "L1M4", name)) %>%
# mutate(name = ifelse(grepl("L1M5", name), "L1M5", name)) %>%
# mutate(name = ifelse(grepl("L1M6", name), "L1M6", name)) %>%
# mutate(name = ifelse(grepl("L1M7", name), "L1M7", name)) %>%
# mutate(name = ifelse(grepl("L1M8", name), "L1M8", name)) %>%
# mutate(name = ifelse(grepl("L1MA", name), "L1MA", name)) %>%
# mutate(name = ifelse(grepl("L1MB", name), "L1MB", name)) %>%
# mutate(name = ifelse(grepl("L1MC", name), "L1MC", name)) %>%
# mutate(name = ifelse(grepl("L1MD", name), "L1MD", name)) %>%
  mutate(name = ifelse(grepl("L1P", name), "L1P", name)) %>%
  filter(name == "L1P") %>%
  ggplot(aes(seqnames,meth, fill = name))+geom_violin()

l1x
```

# Cluster CGIs based on methylation state: example Xist 
```{r Xist, include=F}
reads <- tabix_mbed(paste0(dat, "/whole_genome/chrX_CpG_methylation.bed"),extcol = "motif",by = "read") 

chr="chrX"
phase_start=72278142
phase_end=72278553

size_sel <- reads %>%
  dplyr::mutate(rlen = end-start) %>%
  filter(rlen >= 50000) %>%
  mutate(motif = "CG")

cgcalls <- mbedByCall(size_sel) %>%
  distinct()


mat <- cgcalls %>%
  filter(start >phase_start ) %>%
  filter(end < phase_end) %>%
  select(c(qname, start, mcall)) %>%
 # filter(qname != "9c4675b0-c96f-4411-ba88-b3b8d185e369") %>%
  spread(start, mcall)

mat <- mat  %>%
column_to_rownames("qname")
  

mat[is.na(mat)] <- -1
model1 <- Mclust(mat,G = 1:2, modelNames = "EII")
summary(model1)
clust <- model1$classification
  
df <- as.data.frame(unlist(clust)) %>%
rownames_to_column() %>%
dplyr::rename("qname" = 1, "clust" = 2)

region_start=72216715
region_end=72283838

calls <- cgcalls %>%
  filter(start > region_start) %>%
  filter(end < region_end)

cpg_runs <- getRuns(calls, maxGap = 100)

all_runs <- merge(cpg_runs,df, by = "qname" )
cpg_runs.ordered <- order_reads_facet(all_runs)



cpg_runs_plot <- cpg_runs.ordered$x %>%
    mutate(m = ifelse(values == 1, "Methylated","Unmethylated")) %>%
    mutate(mod = "CpG")
  

pal <- pal_npg("nrc")(10)
meth_pal <- c(pal[1],pal[2])
#ids <- order(sapply(seq_along(gcruns.list),function(i){length(unique(gcruns.list$qname))}))


g <- ggplot(cpg_runs_plot,aes(xmin = start, xmax = end, ymin = ymin, ymax = ymax))+geom_rect(data = cpg_runs.ordered$bounds, fill = "grey80")+ 
    geom_rect(aes(fill = m)) +scale_fill_manual(name = "State", values = meth_pal)+ theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), panel.spacing = unit(2, "lines"))+theme_classic()+coord_cartesian(xlim=c(region_start, region_end))+facet_wrap(~clust, ncol =1)#+geom_vline(xintercept = phase_start)+geom_vline(xintercept = phase_end)
  print(g)
```

```{r CGI_clust}
# load CG islands
CGI <- import.gff(paste0(dat, "/CGIs/chrx.gff")) 
gffRangedData<-import.gff(paste0(dat, "/liftoff/hg38_genes_t2t_liftoff.gtf")) 
genes <- gffRangedData[gffRangedData$type == "gene" ]


# only keep if overlap a gene
CGI_genes <- ovls(CGI, genes)

# remove if overlap SINE, SVA or LINE
TEs <- c("SINE", "LINE", "LTR")
alu <- read_tsv(paste0(dat, "/annotations/t2t-chm13.20200727_repeatmasker.out.bed") ,col_names= F) %>%
  select(X1, X2, X3, X4, X7, X8) %>%
  dplyr::rename("chr" = X1, "start" = X2, "end" = X3, "name" = X4, "rep_type" = X7, "sub_type" =X8) %>%
  mutate(len = end - start) %>%
  filter(chr == "chrX") %>%
  filter(rep_type %in% TEs) %>%
  mutate(ID = row_number())

alu.gr <- GRanges(alu)

CGI_remove <- subsetByOverlaps(GRanges(CGI_genes), alu.gr)
CGI_genes_filtered <- CGI_genes %>%
  filter(!ID %in% CGI_remove$ID)

# if closer than 100bp merge
distanceToNearest(GRanges(CGI_genes_filtered))


```

