---
title: "`r params$newtitle`"
output: pdf_document

---
```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(Biostrings)
library(GenomicRanges)
library(Rsamtools)
library(Sushi)
library(ggplot2)
library(png)

options(knitr.duplicate.label = 'allow')
```



```{r include=FALSE}
tig=params$tig
cen1=(params$cen1)-10000
cen2=(params$cen2)+10000


bismark <- read_tsv(params$bismark, col_names = F)
cov <- read_tsv(params$cov, col_names = F)
BS <- read.bismark(params$bismark)




repeats <- read_tsv(params$repeats, col_names = F) %>%
  separate(X5, c("class", "family"), sep =  "/") %>%
  mutate("family" = replace_na(family, "Unknown")) %>%
  rename(X4 = "repeat_type") %>%
  mutate(name = ifelse(family == c("centr"), paste0(class,"/",family),class)) %>%
  mutate(name = ifelse(family == c("telo"), paste0(class,"/",family),name)) %>%
  mutate(name = ifelse(family == c("acro"), paste0(class,"/",family),name)) %>%
  mutate(name = ifelse(family == c("Unkown") & class == c("Satellite"), paste0(class,"/",family),name)) %>%
  mutate(name = ifelse(repeat_type == c("CER", "(CATTC)n", "(GAATC)n"), repeat_type, name)) %>%
  mutate(name = ifelse(name == "Satellite/centr", repeat_type, name)) %>%
  mutate(name = ifelse(name == "DNA?", "DNA", name)) %>%
  mutate(name = ifelse(grepl("GSAT", repeat_type), "GSAT", name)) %>%
  mutate(name = ifelse(grepl("HSAT", repeat_type), "HSAT", name)) %>%
  mutate(name = ifelse(name == "Unspecified", "Unknown", name)) %>%
  mutate(name = ifelse(class == c("RC"), paste0(class,"/",family),name)) %>%
  mutate(name = as.factor(name))

#repeats$name = gsub("\\?.*", "", repeats$name)

#repeats <- read_tsv("~/chm13_meth/scripts/censat/tig00000514.fasta_rm.bed", col_names = F) %>%
#  mutate(X7 = as.factor(X7))
```

```{r include = FALSE}
smoothed=500
smooth <-BSmooth(BS, 
                 ns=smoothed)

rawmeth <- getMeth(BS, type = "raw")
getmeth <- getMeth(smooth)
getmeth <- as_tibble(cbind(getmeth, bismark$X2), colnames(c("meth", "coords")))

rawmeth <- as_tibble(cbind(bismark$X1, bismark$X2, rawmeth))
rawmeth <- add_column(rawmeth, d = "meth", .after = 2)

write.table(rawmeth, file = paste0("/uru/Data/Nanopore/Analysis/gmoney/chm13/censat/phase_2/wig/",tig, ".wig"), col.names = F, row.names = F, quote = F, sep = "\t")
colnames(getmeth) <- c("meth", "coord")
```



Methylation data for:  `r params$tig` :  `r print(params$cen1)`  -  `r print(params$cen2)`

```{r fig.width=12, fig.height=6}
plot <- getmeth %>% 
  ggplot(aes(x = (coord), y= meth)) +geom_line()+ labs(x="Genomic coordinates (Mb)", y="Methylation")+theme_bw()+xlim(cen1,cen2)

```


```{r fig.width=12, fig.height=6}
depth <- ggplot(cov, aes(x=X2, y=X4))+geom_line()+ labs(x="Genomic coordinates (Mb)", y="read depth")+theme_bw()+xlim(cen1,cen2)

```

```{r fig.width=12, fig.height=6}



    
repeatColors =c("(CATTC)n" = "#E87C71",
    "(GAATC)n"="#E28455",
    "ALR/Alpha"="#D78C32",
    "BSR/Beta"="#E370AB",
    "CER" = "#CE9334",
    "DNA"="#C19935",
    "GSAT"="#B3A033",
    "HSAT"="#A2A638",
    "LINE"="#8CAC3E",
    "Low_complexity"="#75B042",
    "LSAU"="#54B346",
    "LTR"="#51B756",
    "RC/Helitron"="#53BB73",
    "Retroposon"="#55BE8D",
    "RNA"="#54C0A5",
    "rRNA"="#52BEBB",
    "SAR"="#51BDCE",
    "Satellite/acro"="#4EB8DF",
    "Satellite/telo"="#53B0E3",
    "SATR1"="#5AA5DA",
    "scRNA"="#6B9AD2",
    "Simple_repeat"="#8992C8",
    "SINE"="#9A8AC1",
    "snRNA"="#A885BC",
    "srpRNA"="#B67EB6",
    "SSTI"="#C378B2",
    "SUBTEL2_sat"="#D173AF",
    "tRNA"="#ED72A5",
    "Unknown"="#EF768C")

defaultColor = "#000080"


    



rep_leg <- ggplot(data=repeats, mapping=aes(xmin=(X2),xmax=(X3),ymin=-1,ymax=1, fill = name))+
  geom_rect()  +  theme_void() +theme(legend.position="top") +labs(y="Axis")+xlim(cen1,cen2) +labs(y="Axis") + scale_fill_manual(values = repeatColors, drop = FALSE)

require(gridExtra)
all <- grid.arrange(rep_leg,plot, depth, ncol = 1)


```





