---
title: "Run methyl pipeline"
output: pdf_document

---
```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(Biostrings)
library(ggplot2)
library(png)
library(cowplot)
options(scipen=999)
library(zoo)
options(knitr.duplicate.label = 'allow')
```



```{r include=FALSE}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/final_assembly/figures"

dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/final_assembly"
freq = read_tsv(paste0(dat, "/methylation_frequency.tsv"))
BS = read.bismark(paste0(dat, "/bismark.out"))
bismark <- read_tsv(paste0(dat, "/bismark.out"), col_names = F)

censat = read_tsv(paste0(dat, "/annotations/t2t-chm13.20200727.cenSat_annotation.bed"), col_names = F) %>%
  mutate(name = ifelse(grepl("hsat", X4), "HSAT", X4)) %>%
  mutate(name = ifelse(grepl("ct", X4), "CT", name)) %>%
  mutate(name = ifelse(grepl("bsat", X4), "BSAT", name)) %>%
  mutate(name = ifelse(grepl("hor", X4), "HOR", name)) %>%
  mutate(name = ifelse(grepl("mon", X4), "MON", name)) %>%
  mutate(name = ifelse(grepl("Alu", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("SATR", X4), "SATR", name)) %>%
  mutate(name = ifelse(grepl("ACRO", X4), "ACRO", name)) %>%
  mutate(name = ifelse(grepl("GSATII", X4), "GSATII", name)) %>%
  mutate(name = ifelse(grepl("TAR", X4), "TAR", name)) %>%
  mutate(name = ifelse(grepl("TE", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("MER", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("MST", X4), "MST", name)) %>%
  mutate(name = ifelse(grepl("CER", X4), "CER", name)) %>%
  mutate(name = ifelse(grepl("L1", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("SST", X4), "SST", name)) %>%
  mutate(name = ifelse(grepl("LSAU", X4), "LSAU", name)) %>%
  mutate(name = ifelse(grepl("GSAT", X4), "GSAT", name)) %>%
  mutate(name = ifelse(grepl("MSAT", X4), "MSAT", name)) %>%
  mutate(name = ifelse(grepl("novel", X4), "novel", name)) %>%
  mutate(name = ifelse(grepl("HERV", X4), "TE", name)) %>%
  mutate(name = ifelse(grepl("LT", X4), "TE", name)) %>%
  dplyr::select(c(X1, X2, X3, name)) %>%
  dplyr::rename("chrom" =1, "start" = 2 ,"end" =3) %>%
  group_by(name) %>%
  filter(n() >= 3) %>%
  ungroup()

reps <- read_tsv(paste0(dat, "/annotations/t2t-chm13.20200727_repeatmasker.out.bed") ,col_names= F)
                  
repeatColors =c("(CATTC)n" = "#E87C71",
    "(GAATC)n"="#E28455",
    "HOR"="#D78C32",
    "BSAT"="#E370AB",
    "CER" = "#CE9334",
    "DNA"="#C19935",
    "HSAT"="#A2A638",
    "LINE"="#8CAC3E",
    "Low_complexity"="#75B042",
    "LSAU"="#54B346",
    "LTR"="#51B756",
    "MST"="#53BB73",
    "GSAT"="#55BE8D",
    "RNA"="#54C0A5",
    "rRNA"="#52BEBB",
    "SAR"="#51BDCE",
    "ACRO"="#4EB8DF",
    "Satellite/telo"="#53B0E3",
    "SATR"="#5AA5DA",
    "CT"="#6B9AD2",
    "Simple_repeat"="#8992C8",
    "SINE"="#9A8AC1",
    "MON"="#A885BC",
    "SST"="#C378B2",
    "tRNA"="#ED72A5",
    "Unknown"="#EF768C", 
    "gap-rDNA"="#ff4000",
    "TE" = "#ffbf00", 
   "TAR"= "#0080ff" )
```

```{r include = FALSE}
#chr8    42649237        48722262
# smooth to 500 nucleotides and save rds file for easy loading
smooth <-BSmooth(BS, 
                 ns=200)
#saveRDS(smooth, file = paste0(dat, "/chm13_final_CpG_bssmooth_n500.rds"))

#smooth <- readRDS(paste0(dat, "/chm13_final_CpG_bssmooth_n500.rds"))

cov <- getCoverage(smooth)
getmeth <- getMeth(smooth)
getmeth <- dplyr::as_tibble(cbind(getmeth, bismark$X2, bismark$X1, cov))%>%
  dplyr::rename("meth" = 1, "coord" = 2, "chrom" = 3, "cov" = 4) %>%
  mutate("meth" = as.numeric(meth)) %>%
  mutate("coord" = as.numeric(coord)) %>%
  mutate("chrom" = as.factor(chrom))

```

```{r grangesobj, include=F}

table(censat$name)
censat.gr <- GRanges(as.data.frame(censat))

freq_sub <- freq %>%
  dplyr::select("chromosome", "start", "end", "methylated_frequency") %>%
  rename(chromosome = "chrom")
freq.gr <- GRanges(freq_sub)

keepi <- findOverlaps(freq.gr,censat.gr)
freq.matched <- freq.gr[queryHits(keepi)]

mcols(freq.matched) <- cbind.data.frame(
    mcols(freq.matched),
    mcols(censat.gr[subjectHits(keepi)]))

freq.matched

censat_meth <- as.data.frame(freq.matched)
```


```{r fig.width=12, fig.height=6}

cen <- read_tsv(paste0(dat, "/annotations/t2t-chm13.20200727.cenRegions.bed"), col_names = F)

for (i in 1:length(cen$X1)){
  

chr=cen$X1[i]
start=cen$X2[i]
end=cen$X3[i]

cpgs <- getmeth %>%
  filter(chrom == chr) %>%
  filter(coord >= start) %>%
  filter(coord <= end)
  


plot <- cpgs %>% 
  ggplot(aes(x = (coord/1e6), y= meth)) +geom_line()+ labs(x="Genomic coordinates (Mb)", y="Methylation")+theme_classic(base_size = 25)+xlim(start/1e6,end/1e6)+ylim(0,1)



defaultColor = "#000080"


censat_reg <- censat %>%
  filter(chrom == chr) 

rep_leg <- ggplot(data=censat_reg, mapping=aes(xmin=(start),xmax=(end),ymin=0,ymax=.1, fill = name))+
  geom_rect()  +  theme_void() +theme(legend.position="top") +labs(y="Axis")+xlim(start,end) +labs(y="Axis")+ scale_fill_manual(values = repeatColors, drop = FALSE)+theme(legend.text=element_text(size=rel(1.5)))+theme(legend.title=element_blank())


cov <- cpgs %>% 
  mutate(cov = as.numeric(cov)) %>%
  arrange(coord) %>%
  mutate(smooth = rollmean(cov, 1000, fill = NA)) %>%
  ggplot(aes(x = (coord/1e6), y= smooth))+geom_line(size =1) + labs(x="Genomic coordinates (Mb)", y="Methylation")+theme_classic(base_size = 25)+xlim(start/1e6,end/1e6)




top_row <- plot_grid(rep_leg,plot, cov, ncol = 1, align="v")

ggsave(
  paste0(figs, "/", chr, "_methyl_pattern.pdf"),
  plot = top_row,
  scale = 1,
  width = 12,
  height = 12,
)

censat_meth <- as.data.frame(freq.matched) %>%
  filter(seqnames == chr)

violin <- ggplot(data = censat_meth, aes(factor(name), methylated_frequency, fill = name))+geom_violin()+theme_classic(base_size = 20)+ scale_fill_manual(values = repeatColors, drop = FALSE)+labs(x = "Repeat type", y = "Methylation frequency")

ggsave(
  paste0(figs, "/", chr, "_methyl_quant.pdf"),
  plot = violin,
  scale = 1,
  width = 10,
  height = 6,
)

}

```
