---
title: "draft_v7 chm13 chromosome X methylation"
output: pdf_document

---

```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(Biostrings)
library(GenomicRanges)
library(Rsamtools)
library(Sushi)
library(ggplot2)
library(png)

options(knitr.duplicate.label = 'allow')
```

## Methylation report

```{r include=FALSE}
path <- "/uru/Data/Nanopore/Analysis/gmoney/chm13/draft_v0.7/"

freq <- read_tsv(paste0(path, "methylation_frequency.tsv"))
mcalls <- read_tsv(paste0(path, "methylation_calls.tsv"))
bismark <- read_tsv(paste0(path, "meth_call/", "bismark.out"), col_names = F)
cov <- read_tsv(paste0(path, "meth_call/","chm13_meth.chrX.bg"), col_names = F)

# parse files for the given roi
chr <- "chrX_v0.7"
coord1 <- 57828561
coord2 <- 60934693

smoothed = 500

# calculate percent methylated and percent of high quality calls

m <- length(which(mcalls$log_lik_ratio>2.5))
u <- length(which(mcalls$log_lik_ratio<(-2.5)))
a <- length(mcalls$log_lik_ratio)

calls <- round((((m+u)/a)*100), digits = 2)
meth <- round(((m/(m+u))*100), digits=2)

```

```{r include = FALSE}
mcalls_roi <- mcalls %>% 
  filter(chromosome == chr) %>%
  filter(start >= coord1) %>%
  filter(end <= coord2)

m <- length(which(mcalls_roi$log_lik_ratio>2.5))
u <- length(which(mcalls_roi$log_lik_ratio<(-2.5)))
a <- length(mcalls_roi$log_lik_ratio)

calls <- round((((m+u)/a)*100), digits = 2)
meth <- round(((m/(m+u))*100), digits=2)
```
Centromere methylation Summary stats

Coordinates: chrX:57828561 â€“ 60934693

Percent high quality calls: `r calls`%  

Percent methylation: `r meth`%  

Histogram of log likelihood ratios

``` {r}
print(ggplot(data = mcalls_roi, aes(x=log_lik_ratio))+geom_histogram(alpha=.5, binwidth = .5, fill = "red")+theme_bw()+geom_vline(xintercept= c(-2.5, 2.5), linetype="dashed", color="black")+xlim(-25,25))

```
\newpage
Methylation frequency per base

```{r include = FALSE}
freq$pos <- rep("chrom", nrow(freq))
freq[freq$start >= coord1 & freq$end  <= coord2 , ][, "pos"] <- "cen"
```

``` {r}
print(ggplot(data = freq, aes(x = methylated_frequency, fill=freq$pos)) +geom_density(alpha=.5)+facet_wrap(~freq$pos, ncol=2)+theme_bw()+labs(x="methylated frequency (per base)", color="Position")+scale_fill_manual("region",values=c( "red", "green"))+theme(legend.position="bottom"))

```
\newpage
Methylation frequency 50bp bins

``` {r}
print(binned <- freq %>%
  mutate(bin = ntile(start, (length(freq$start)/50))) %>%
  group_by(bin, pos) %>%
  summarise(avgmeth = mean(methylated_frequency)) %>%
  arrange(desc(avgmeth)) %>% ggplot(aes(x=avgmeth, fill = pos))+geom_density(alpha=.5)+theme_bw()+labs(x="Average methylation frequency (bin size 50bps)")+facet_wrap(~pos, ncol=2)+scale_fill_manual("Region",values=c( "red", "green"))+theme(legend.position="bottom"))
```

\newpage

```{r include = FALSE}
BS <- read.bismark(paste0(path, "meth_call/", "bismark.out"))
smooth <-BSmooth(BS, 
                 ns=smoothed)
cenx <- data.frame(start = coord1, end = coord2, chr = chr)
#chr8_v0.1_unpolished:44,825,722-44,902,661
#roi <- data.frame(start=44825722,end=44902661, chr="chr8_v0.1_unpolished")
getmeth <- getMeth(smooth)
getmeth <- as_tibble(cbind(getmeth, bismark$X2))
colnames(getmeth) <- c("meth", "coord")
```

Methylation plot

smooth: $`r smoothed`$ bp

```{r fig.width=12, fig.height=6}
wide = 500
plot <- getmeth %>% 
  filter(coord > (coord1)) %>%
  filter(coord < (coord2)) %>%
  ggplot(aes(x = (coord/1000), y= meth)) +geom_line()+ labs(x="Genomic coordinates (Kb)", y="Methylation")+theme_bw()

print(plot)

ggsave(filename = "/uru/Data/Nanopore/Analysis/gmoney/chm13/draft_v0.7/meth_call/figures/chrx_methylation.pdf", plot = plot, width = 12, height = 6)

```


Coverage plot

```{r fig.width=12, fig.height=6}
#pdf(file = "/uru/Data/Nanopore/Analysis/gmoney/chm13/draft_v0.7/meth_call/figures/cenx_coverage.pdf",
#    width =15,
#    height = 8)

plotBedgraph((as.data.frame(cov)),chr,(coord1-wide),(coord2+wide), transparency =.5, color= SushiColors(2)(2)[1])
labelgenome("cenX",coord1,coord2,n=3,scale="Kb")
mtext("Read Depth",side=2,line=1.75,cex=1,font=2)
axis(side=2,las=2,tcl=.2)


```


\newpage

IGV image of chrX:59,213,083-59,306,271 (hypomethylated region in the middle of cenX)
```{r echo=FALSE,out.width = '100%'}
# save igv image as png and input path
knitr::include_graphics("/uru/Data/Nanopore/Analysis/gmoney/chm13/draft_v0.7/meth_call/figures/chrx_middle.png")
```



