---
title: "Nanonome Read Binning"
author: "G$"
date: "9/3/2020"
output: pdf_document
---

---
```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(Biostrings)
library(ggplot2)
library(png)
library(cowplot)
options(scipen=999)
library(zoo)
library(ggpmisc)
library(Repitools)
options(knitr.duplicate.label = 'allow')
source("/home/isac/Code/ilee/plot/ilee_plot_utils.R")
library("ggsci")
source("/home/isac/Code/nanopore-methylation-utilities/methylation_R_utils.R")
```

```{r mbed}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/nanonome/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/HG002/nanonome/methylation_calls"

```


```{r functions, include = F}
getQQ <- function(one,two,qs = seq(0,1,0.01)) {
  tibble(one = quantile(one,qs,na.rm = T),
         two = quantile(two,qs,na.rm = T),
         qunatile = qs)
}
# https://stackoverflow.com/questions/43875716/find-start-and-end-positions-indices-of-runs-consecutive-values
#getRuns <- function(calls){
#  calls <- calls %>%
#    filter(mcall != -1) %>%
#    group_by(qname)
#  calls.list <- calls %>%
#    group_split(keep = F)
#  names(calls.list) <- calls %>% group_keys() %>% .$qname
#  runs.list <- lapply(calls.list,function(x){
#    rle(x$mcall) %>%
#    unclass() %>% as.tibble() %>%
#    mutate( endi = cumsum(lengths),
#            starti = c(1,dplyr::lag(endi)[-1]+1),
#            start = x$start[starti],
#            end = x$start[endi],
#            width = end - start + 1) %>%
#      filter( width >= 0) # remove negative widths (in case of dups, etc.)
#  })
#  runs <- bind_rows(runs.list,.id = "qname")
#}
order_reads <- function(x,bounds=NULL){
  # get boundaries of reads if not provided
  if (is.null(bounds)){
    bounds <- x%>% group_by(qname) %>%
      summarize(start = min(start),
                end = max(end))
    # label y based on order of smallest start
    bounds<- bounds %>% arrange(start, end) %>%
      mutate(
        readi = seq_len(length(unique(x$qname))),
        ymin = -readi - 0.8, 
        ymax = ymin + 0.6)
  }
  x <- x %>%
    mutate(ymin = bounds$ymin[match(qname,bounds$qname)],
           ymax = bounds$ymax[match(qname,bounds$qname)])
  bounds <- bounds %>%
    mutate(ymin = bounds$ymin[match(qname,bounds$qname)],
           ymax = bounds$ymax[match(qname,bounds$qname)])
  return(list(x = x,bounds = bounds))
}

smoothCalls <- function(calls,reg=NULL,bandwidth = 1){
  calls <- calls %>%
    mutate(mcall = ifelse(abs(score)>1,sign(score),score)) # ceiling based on log-lik ratio - this uses log-lik ratio when call is not obvious
  if (is.null(reg)) {
    xpoints <- seq(min(calls$start),max(calls$start))
  } else {
    reg <- as_tibble(reg)
    xpoints <- seq(reg$start,reg$end)
  }
  ks <- ksmooth(calls$start,calls$mcall,bandwidth = 1,kernel = "normal",x.points = xpoints)
  tibble(
    start = ks$x,
    mcall_smooth = ks$y, 
    mcall = case_when(
      mcall_smooth > 0 ~ 1,
      mcall_smooth < 0 ~ 0,
      TRUE ~ -1)) 
}

binnedSum <- function(bins, numvar, mcolname)
{
  stopifnot(is(bins, "GRanges"))
  stopifnot(is(numvar, "RleList"))
  stopifnot(identical(seqlevels(bins), names(numvar)))
  bins_per_chrom <- split(ranges(bins), seqnames(bins))
  sums_list <- lapply(names(numvar),
      function(seqname) {
          views <- Views(numvar[[seqname]],
                         bins_per_chrom[[seqname]])
          viewSums(views)
      })
  new_mcol <- unsplit(sums_list, as.factor(seqnames(bins)))
  mcols(bins)[[mcolname]] <- new_mcol
  bins
}
```


```{r load_data, include=F}
readsGC <- tabix_mbed(paste0(dat, "/chrX:52000000-63000000_GpCmethylation.merge1-3.bed"),extcol = "motif",by = "read") 

size_sel <- readsGC %>%
  mutate(rlen = end-start) %>%
  filter(rlen >= 20000) %>%
  mutate(motif = "GC")

gccalls <- mbedByCall(size_sel) %>%
  drop_na(mcall)

reads <- tabix_mbed(paste0(dat, "/chrX:52000000-63000000_CpGmethylation.merge1-3.bed"),extcol = "motif",by = "read") 
#chrX:48,626,035-49,040,073



size_sel <- reads %>%
  mutate(rlen = end-start) %>%
  filter(rlen >= 20000) %>%
  mutate(motif = "CG")



cgcalls <- mbedByCall(size_sel) %>%
  drop_na(mcall)
```


```{r bin_GC, echo=F}

cgsmooth.list <- gccalls %>%
  select(qname, start, mcall) %>%
  mutate(mcall_smooth = case_when(
      mcall > 0 ~ 1,
      mcall < 0 ~ 0,
      TRUE ~ -1)) 

runs <- getRuns(cgsmooth.list)

runs <- runs %>%
  mutate(color = case_when(values == 0 ~ "Inaccessible", 
                           TRUE ~ "Accessible"))

ggplot(runs, aes(width))+geom_density()+xlim(0,1000)+facet_wrap(~color, scales = "free")
```

```{r gethypo, echo=F}

cgcalls.gr <-  cgcalls %>%
  ungroup() %>%
  group_by(start) %>%
  summarise(num_meth = sum(mcall == 1), num_unmeth = sum(mcall == 0)) %>%
  mutate(chr = "chrX", start = start, end = start) %>%
  GRanges()

blocks <- genomeBlocks(BSgenome.HG002.chrX, chrs = seqnames(BSgenome.HG002.chrX), width = 5000)
score1 <- coverage(cgcalls.gr, weight="num_meth")
score2 <- coverage(cgcalls.gr, weight="num_unmeth")

binned_meth <- binnedSum(blocks, numvar = score1, "num_meth") %>%
  as.data.frame()
binned_unmeth <-binnedSum(blocks, numvar = score2, "num_unmeth")%>%
  as.data.frame()

meth_bins <- merge(binned_meth, binned_unmeth, by = c("start", "end", "seqnames", "width", "strand")) %>%
  filter(start > 55920000) %>%
  filter(end < 59100000) %>%
  group_by(start, end) %>%
  mutate(sites = num_meth+num_unmeth) %>%
  mutate(freq = num_meth/sites) %>%
  na.omit()


meth_CG <- ggplot(meth_bins, aes(x = (start/1e6), y= freq)) + labs(x="Genomic coordinates (Mb)", y="Methylation")+theme_classic(base_size = 25)+ylim(0,1)+geom_line(size =1)+geom_hline(yintercept = .40, linetype = "dashed")
meth_CG
```

```{r}
runs_regs <- runs %>%
  mutate(reg = case_when(start >= 55920000 & end <= 59100000 ~ "alphaSat", 
                   TRUE ~ "CT")) %>%
  filter(start > 53000000) %>%
  filter(end < 59100000) %>%
  filter(color == "Inaccessible")

ggplot(runs_regs, aes(width))+geom_histogram(aes(y=..density..,fill = reg, alpha=.2),binwidth=4, position="identity")+facet_wrap(~color+reg, scales = "free")+xlim(0,1000)+geom_density(aes(y=..density.., alpha=.2), adjust = .75)+theme_classic()
```


```{r}

asat_closed <- runs_regs %>%
  filter(reg=="alphaSat") %>%
  filter(color == "Inaccessible")

p <- ggplot(asat_closed, aes(width,alpha = .5))+geom_density(adjust = .85)+xlim(0,1000)

pb <- ggplot_build(p)
p + stat_peaks(
  data = pb[['data']][[1]], # take a look at this object
  aes(x = x, y = density),
  colour = "red",
  geom = "text",
  size = 3
)

CT_closed <- runs_regs %>%
  filter(reg=="CT") %>%
  filter(color == "Inaccessible")

p <- ggplot(CT_closed, aes(width,alpha = .5))+geom_density(adjust=.75)+xlim(0,1000)

pb <- ggplot_build(p)
p + stat_peaks(
  data = pb[['data']][[1]], # take a look at this object
  aes(x = x, y = density),
  colour = "red",
  geom = "text",
  size = 3
)

```

