---
title: "Nanonome Read Binning"
author: "G$"
date: "9/3/2020"
output: pdf_document
---

---
```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(Biostrings)
library(ggplot2)
library(png)
library(cowplot)
options(scipen=999)
library(zoo)
library(ggpmisc)
library(Repitools)
options(knitr.duplicate.label = 'allow')
source("/home/isac/Code/ilee/plot/ilee_plot_utils.R")
library("ggsci")
source("/home/isac/Code/nanopore-methylation-utilities/methylation_R_utils.R")
```

```{r mbed}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/nanonome/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/HG002/nanonome/methylation_calls"

```


```{r functions, include = F}
getQQ <- function(one,two,qs = seq(0,1,0.01)) {
  tibble(one = quantile(one,qs,na.rm = T),
         two = quantile(two,qs,na.rm = T),
         qunatile = qs)
}
# https://stackoverflow.com/questions/43875716/find-start-and-end-positions-indices-of-runs-consecutive-values
#getRuns <- function(calls){
#  calls <- calls %>%
#    filter(mcall != -1) %>%
#    group_by(qname)
#  calls.list <- calls %>%
#    group_split(keep = F)
#  names(calls.list) <- calls %>% group_keys() %>% .$qname
#  runs.list <- lapply(calls.list,function(x){
#    rle(x$mcall) %>%
#    unclass() %>% as.tibble() %>%
#    mutate( endi = cumsum(lengths),
#            starti = c(1,dplyr::lag(endi)[-1]+1),
#            start = x$start[starti],
#            end = x$start[endi],
#            width = end - start + 1) %>%
#      filter( width >= 0) # remove negative widths (in case of dups, etc.)
#  })
#  runs <- bind_rows(runs.list,.id = "qname")
#}
order_reads <- function(x,bounds=NULL){
  # get boundaries of reads if not provided
  if (is.null(bounds)){
    bounds <- x%>% group_by(qname) %>%
      summarize(start = min(start),
                end = max(end))
    # label y based on order of smallest start
    bounds<- bounds %>% arrange(start, end) %>%
      mutate(
        readi = seq_len(length(unique(x$qname))),
        ymin = -readi - 0.8, 
        ymax = ymin + 0.6)
  }
  x <- x %>%
    mutate(ymin = bounds$ymin[match(qname,bounds$qname)],
           ymax = bounds$ymax[match(qname,bounds$qname)])
  bounds <- bounds %>%
    mutate(ymin = bounds$ymin[match(qname,bounds$qname)],
           ymax = bounds$ymax[match(qname,bounds$qname)])
  return(list(x = x,bounds = bounds))
}

smoothCalls <- function(calls,reg=NULL,bandwidth = 1){
  calls <- calls %>%
    mutate(mcall = ifelse(abs(score)>1,sign(score),score)) # ceiling based on log-lik ratio - this uses log-lik ratio when call is not obvious
  if (is.null(reg)) {
    xpoints <- seq(min(calls$start),max(calls$start))
  } else {
    reg <- as_tibble(reg)
    xpoints <- seq(reg$start,reg$end)
  }
  ks <- ksmooth(calls$start,calls$mcall,bandwidth = 1,kernel = "normal",x.points = xpoints)
  tibble(
    start = ks$x,
    mcall_smooth = ks$y, 
    mcall = case_when(
      mcall_smooth > 0 ~ 1,
      mcall_smooth < 0 ~ 0,
      TRUE ~ -1)) 
}

binnedSum <- function(bins, numvar, mcolname)
{
  stopifnot(is(bins, "GRanges"))
  stopifnot(is(numvar, "RleList"))
  stopifnot(identical(seqlevels(bins), names(numvar)))
  bins_per_chrom <- split(ranges(bins), seqnames(bins))
  sums_list <- lapply(names(numvar),
      function(seqname) {
          views <- Views(numvar[[seqname]],
                         bins_per_chrom[[seqname]])
          viewSums(views)
      })
  new_mcol <- unsplit(sums_list, as.factor(seqnames(bins)))
  mcols(bins)[[mcolname]] <- new_mcol
  bins
}
```


```{r load_data, include=F}
readsGC <- tabix_mbed(paste0(dat, "/unpooled/HG002_nanonome4-5_winnowmapk15_chrM_GpCmethylation.bed.gz"),extcol = "motif",by = "read") 

size_sel <- readsGC %>%
  mutate(rlen = end-start) %>%
#  filter(rlen >= 20000) %>%
  mutate(motif = "GC")

gccalls <- mbedByCall(size_sel) %>%
  drop_na(mcall)

```


```{r bin_GC, echo=F}

cgsmooth.list <- gccalls %>%
  select(qname, start, mcall) %>%
  mutate(mcall_smooth = case_when(
      mcall > 0 ~ 1,
      mcall < 0 ~ 0,
      TRUE ~ -1)) 

runs <- getRuns(cgsmooth.list)

runs <- runs %>%
  mutate(color = case_when(values == 0 ~ "Inaccessible", 
                           TRUE ~ "Accessible")) %>%
  filter(color == "Inaccessible")

ggplot(runs, aes(width))+geom_histogram(binwidth = 5)+xlim(0,1000)

ggsave(
  paste0(figs, "/ChrM_runs_histogram.pdf"),
  plot = last_plot(),
  width = 8,
  height = 6
)

```


```{r load_data, include=F}

gccalls <- read_csv(paste0(dat, "/pooled/HG002_nanonome_chrX_GpCmethylation_ByCall.csv"))

```


```{r bin_GC, echo=F}

cgsmooth.list <- gccalls %>%
  select(qname, start, mcall) %>%
  mutate(mcall_smooth = case_when(
      mcall > 0 ~ 1,
      mcall < 0 ~ 0,
      TRUE ~ -1)) 

runs <- getRuns(cgsmooth.list)

runs <- runs %>%
  mutate(color = case_when(values == 0 ~ "Inaccessible", 
                           TRUE ~ "Accessible"))
#saveRDS(runs, paste0(dat, "/pooled/HG002_nanonome_chrX_GpC_runsUnsmoothed.rds"))
ggplot(runs, aes(width))+geom_histogram(binwidth = 5)+xlim(0,1000)+facet_wrap(~color, scales = "free")

ggsave(
  paste0(figs, "/ChrX_runs_histogram.pdf"),
  plot = last_plot(),
  width = 8,
  height = 6
)
```
```{r}
runs <- readRDS(paste0(dat, "/pooled/HG002_nanonome_chrX_GpC_runsUnsmoothed.rds"))
```

```{r}

access <- runs %>%
  filter(color == "Accessible")
runs.reg1 <- access %>%
  filter(start > 55483765) %>%
  filter(end < 59373489) %>%
  mutate(reg = "cen")

runs.reg2 <- access %>%
  filter(start > 60000000) %>%
  filter(end < 63000000)%>%
  mutate(reg = "CT")

runs.reg3 <- access %>%
  filter(start > 112297349) %>%
  filter(end < 112514845)%>%
  mutate(reg = "DXZ4") 

runs_all <- rbind(runs.reg1, runs.reg2, runs.reg3)

ggplot(runs_all, aes(width, alpha = .2))+xlim(0,150)+facet_wrap(~color+reg, scales = "free")+geom_histogram(aes(y=..density..,fill = reg), binwidth = 2)+geom_density(adjust=.95)
#+geom_density(adjust=2)

#ggsave(
#  paste0(figs, "/cen_DXZ4_runs_histogram_noSmooth.pdf"),
#  plot = last_plot(),
#  width = 10,
#  height = 10
#)

```


```{r, echo=F, fig.height=3, fig.width=9}

access <- runs %>%
  filter(color == "Inaccessible")
runs.reg1 <- access %>%
  filter(start > 55483765) %>%
  filter(end < 59373489) %>%
  mutate(reg = "cen")

runs.reg2 <- access %>%
  filter(start > 60000000) %>%
  filter(end < 63000000)%>%
  mutate(reg = "CT")

runs.reg3 <- access %>%
  filter(start > 112297349) %>%
  filter(end < 112514845)%>%
  mutate(reg = "DXZ4") 

runs_all <- rbind(runs.reg1, runs.reg2, runs.reg3)

ggplot(runs_all, aes(width, alpha = .2))+xlim(0,1000)+facet_wrap(~color+reg, scales = "free")+geom_histogram(aes(y=..density..,fill = reg), binwidth = 5)+geom_density(adjust=.95)
#+geom_density(adjust=2)

ggplot(runs_all, aes(width, alpha = .2))+xlim(0,1000)+facet_wrap(~color+reg, scales = "free")+geom_histogram(aes(fill = reg), binwidth = 5)
#+geom_density(adjust=2)

#ggsave(
#  paste0(figs, "/cen_DXZ4_runs_histogram_noSmooth.pdf"),
#  plot = last_plot(),
#  width = 10,
#  height = 10
#)

```
